# Классификация комментариев

Интернет-магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Требуется обучить модель классифицировать комментарии на позитивные и негативные. В нашем распоряжении набор данных с разметкой о токсичности правок.

Требуется построить модель со значением метрики качества *F1* не меньше 0.75. 

<h1>Содержание<span class="tocSkip"></span></h1>
</li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

## Подготовка

In [1]:
import pandas as pd
import numpy as np
import torch
import transformers
from tqdm import notebook
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingClassifier
from catboost import CatBoostClassifier
from sklearn.model_selection import cross_val_score, train_test_split 
from sklearn.metrics import make_scorer, f1_score
from sklearn.model_selection import GridSearchCV
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.stem import SnowballStemmer

In [2]:
data = pd.read_csv('datasets/toxic_comments.csv')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


Оценим баланс классов в исходном датасете.

In [6]:
data['toxic'].value_counts()[1]/data['toxic'].value_counts()[0]

0.1131876717871444

Баланс классов меньше 0.5, поэтому при разделении на выборки будем использовать стратификацию.

Инициализируем токенизатор:

In [7]:
tokenizer = transformers.BertTokenizer.from_pretrained('bert-base-uncased')

Проверим работу:

In [8]:
tokenizer.encode('Hello world', add_special_tokens=True) 

[101, 7592, 2088, 102]

Токенизируем слова в `data['text']`:

In [9]:
tokenized = data['text'].apply(
    lambda x: tokenizer.encode(x, add_special_tokens=True))
tokenized

Token indices sequence length is longer than the specified maximum sequence length for this model (631 > 512). Running this sequence through the model will result in indexing errors


0         [101, 7526, 2339, 1996, 10086, 2015, 2081, 210...
1         [101, 1040, 1005, 22091, 2860, 999, 2002, 3503...
2         [101, 4931, 2158, 1010, 1045, 1005, 1049, 2428...
3         [101, 1000, 2062, 1045, 2064, 1005, 1056, 2191...
4         [101, 2017, 1010, 2909, 1010, 2024, 2026, 5394...
                                ...                        
159566    [101, 1000, 1024, 1024, 1024, 1024, 1024, 1998...
159567    [101, 2017, 2323, 2022, 14984, 1997, 4426, 200...
159568    [101, 13183, 6290, 26114, 1010, 2045, 2015, 20...
159569    [101, 1998, 2009, 3504, 2066, 2009, 2001, 2941...
159570    [101, 1000, 1998, 1012, 1012, 1012, 1045, 2428...
Name: text, Length: 159571, dtype: object

Посмотрим, какова максимальная длина строки.

In [10]:
def maxlen(df):
    max_len = 0
    for i in df.values:
        if len(i) > max_len:
            max_len = len(i)
    return max_len

In [11]:
maxlen(tokenized)

4950

Токенов в строке многовато для BERT, а наш датасет слишком велик, и попытка построени embeddings на всём датасете на моём ноутбуке с 16 Гб оперативной памяти ориентировочно занимает 75 часов. Уменьшим датасет, чтобы сэкономить время.

Чтобы модель BERT могла работать, оставим только строки, не превышающие 512 токенов.

In [12]:
tokenized512 = tokenized[tokenized.map(len) <= 512]
tokenized512

0         [101, 7526, 2339, 1996, 10086, 2015, 2081, 210...
1         [101, 1040, 1005, 22091, 2860, 999, 2002, 3503...
2         [101, 4931, 2158, 1010, 1045, 1005, 1049, 2428...
3         [101, 1000, 2062, 1045, 2064, 1005, 1056, 2191...
4         [101, 2017, 1010, 2909, 1010, 2024, 2026, 5394...
                                ...                        
159566    [101, 1000, 1024, 1024, 1024, 1024, 1024, 1998...
159567    [101, 2017, 2323, 2022, 14984, 1997, 4426, 200...
159568    [101, 13183, 6290, 26114, 1010, 2045, 2015, 20...
159569    [101, 1998, 2009, 3504, 2066, 2009, 2001, 2941...
159570    [101, 1000, 1998, 1012, 1012, 1012, 1045, 2428...
Name: text, Length: 156048, dtype: object

Убедимся в том, что максимальная длина строки не превышает 512 токенов.

In [13]:
max_len = maxlen(tokenized512)
max_len

512

Сделаем выборку.

In [14]:
tokenized512 = tokenized512.sample(6000, random_state=4)
tokenized512

79345     [101, 9268, 5310, 5857, 2017, 2079, 2025, 6366...
56195     [101, 1000, 1024, 1024, 1024, 2045, 2003, 2023...
88061     [101, 4512, 2007, 2033, 3849, 2066, 2017, 9471...
117948    [101, 2045, 1005, 1055, 7564, 1997, 6971, 1394...
89217     [101, 9610, 13327, 2846, 2045, 2323, 2022, 103...
                                ...                        
63806     [101, 1000, 3531, 2079, 2025, 3158, 9305, 4697...
92596     [101, 1045, 2903, 2057, 2323, 5787, 11619, 199...
74153     [101, 2748, 1045, 1042, 1008, 23616, 2075, 207...
103249       [101, 1012, 2009, 2003, 2525, 1999, 2486, 102]
68436     [101, 1000, 1024, 2004, 2000, 2115, 3160, 1024...
Name: text, Length: 6000, dtype: object

Применим метод padding (англ. «отступ»), чтобы после токенизации длины исходных текстов в корпусе были равными. Только при таком условии будет работать модель BERT. Пусть стандартной длиной вектора n будет длина наибольшего во всём датасете вектора. Остальные векторы дополним нулями:

In [15]:
padded = np.array([i + [0]*(max_len - len(i)) for i in tokenized512.values])
padded

array([[ 101, 9268, 5310, ...,    0,    0,    0],
       [ 101, 1000, 1024, ...,    0,    0,    0],
       [ 101, 4512, 2007, ...,    0,    0,    0],
       ...,
       [ 101, 2748, 1045, ...,    0,    0,    0],
       [ 101, 1012, 2009, ...,    0,    0,    0],
       [ 101, 1000, 1024, ...,    0,    0,    0]])

Теперь поясним модели, что нули не несут значимой информации. Это нужно для компоненты модели, которая называется «внимание» (англ. attention). Отбросим эти токены и «создадим маску» для действительно важных токенов, то есть укажем нулевые и не нулевые значения:

In [16]:
attention_mask = np.where(padded != 0, 1, 0)
attention_mask

array([[1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       ...,
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0]])

Инициализируем конфигурацию BertConfig.

In [17]:
config = transformers.BertConfig.from_pretrained('bert-base-uncased')

Инициализируем саму модель класса BertModel. Передадим ей файл с предобученной моделью и конфигурацией:

In [18]:
model = transformers.BertModel.from_pretrained('bert-base-uncased', config=config)

In [19]:
#data

In [20]:
#torch.LongTensor(padded[500:1000])

In [21]:
batch_size = 500
embeddings = []
for i in notebook.tqdm(range(padded.shape[0] // batch_size)):
    """
    Преобразуем данные в формат тензоров (англ. tensor) — многомерных векторов 
    в библиотеке torch. Тип данных LongTensor (англ. «длинный тензор») хранит 
    числа в «длинном формате», то есть выделяет на каждое число 64 бита.
    """
    batch = torch.LongTensor(padded[batch_size*i:batch_size*(i+1)])
    # преобразуем маску
    attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)])
    """
    Для ускорения вычисления функцией no_grad() (англ. no gradient, «нет градиента») 
    в библиотеке torch укажем, что градиенты не нужны: модель BERT обучать не будем.
    """
    with torch.no_grad():
        #Чтобы получить эмбеддинги для батча, передадим модели данные и маску:
        batch_embeddings = model(batch, attention_mask=attention_mask_batch)
    """
    Из полученного тензора извлечём нужные элементы и добавим в список всех эмбеддингов:
    """
    embeddings.append(batch_embeddings[0][:,0,:].numpy())

Соберём все эмбеддинги в матрицу признаков:

In [22]:
features = np.concatenate(embeddings)
len(features)

6000

In [23]:
target = data.loc[tokenized512.index]['toxic']
len(target)

6000

Мы подготовили признаки и целевой признак.

## Обучение

Разобьём датасет на обучающую и тестовую выборки.

In [25]:
features_train, features_test, target_train, target_test = \
train_test_split(features, target, test_size = 0.2, stratify=target, random_state=12345)

In [28]:
# Функция возвращает предсказания лучшей модели
def best_model(features_train, 
                  target_train,
                  model,
                  param_grid, 
                  cv, 
                  refit):
    
    my_scorer = {'F1': make_scorer(f1_score)
                #,'MAE': make_scorer(mean_absolute_error)
                }
    
    model = model
    grid = GridSearchCV(model, 
                        param_grid, 
                        cv=cv, 
                        refit=refit, 
                        scoring=my_scorer) 
    grid_result = grid.fit(features_train, target_train)
    best_model = grid_result.best_estimator_
    best_score = grid_result.best_score_
    best_params = grid.best_params_
    print(best_params)
    #print(refit, ": {:.2f}".format(best_score), "%")
    print(refit,":", best_score)
    return(best_model, best_score)

Обучим различные модели.

In [57]:
# Сетка параметров для логистической регрессии
PARAM_GRID_LR = {#'penalty': ['l1', 'l2', 'elasticnet', 'none'],
                 'max_iter': [10000],
                 'class_weight': ['balanced'],
                 'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
                 ,'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
                }
PARAM_GRID_LR;

In [58]:
%%time
model_LR, F1_LR = best_model(features_train, 
                  target_train,
                  LogisticRegression(),
                  PARAM_GRID_LR, 
                  5, 
                  'F1')

C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\optimize.py:204: UserWarning: Line Search failed
  warnings.warn('Line Search failed')
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\optimize.py:204: UserWarning: Line Search failed
  warnings.warn('Line Search failed')
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\

C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\optimize.py:204: UserWarning: Line Search failed
  warnings.warn('Line Search failed')
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\optimize.py:204: UserWarning: Line Search failed
  warnings.warn('Line Search failed')
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\scipy\optimize\linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
C:\Users\shepelev.ao\Anaconda3\lib\site-packages\

{'C': 1, 'class_weight': 'balanced', 'max_iter': 10000, 'solver': 'liblinear'}
F1 : 0.6221613123277205
Wall time: 1h 4min 22s


{'class_weight': 'balanced', 'max_iter': 10000, 'solver': 'liblinear'}
F1 : 0.6221613123277205
Wall time: 2min 23s

In [33]:
# Сетка параметров для DecisionTreeClassifier()
PARAM_GRID_DTC = {'criterion': ['gini', 'entropy'],
                  'splitter': ['best', 'random'],
                  'max_depth': range(1,20),
                  'min_samples_leaf': range(1,10),
                  'random_state': [12345]
                 }  
PARAM_GRID_DTC;

In [34]:
%%time
model_DTC, F1_DTC = best_model(features_train, 
                  target_train,
                  DecisionTreeClassifier(),
                  PARAM_GRID_DTC, 
                  5, 
                  'F1')

{'criterion': 'entropy', 'max_depth': 7, 'min_samples_leaf': 9, 'random_state': 12345, 'splitter': 'random'}
F1 : 0.42693844403029546
Wall time: 1h 14min 14s


{'criterion': 'entropy', 'max_depth': 7, 'min_samples_leaf': 9, 'random_state': 12345, 'splitter': 'random'}
F1 : 0.42693844403029546
Wall time: 1h 14min 14s

In [35]:
# Сетка параметров для RandomForestClassifier()
PARAM_GRID_RFC = {'n_estimators': [3,  10,  31, 100], #np.logspace(0, 2, 5, endpoint=True, dtype = 'int')
                  'max_depth': range(1,5),
                  'min_samples_leaf': range(1,5),
                 'random_state': [12345]}  
PARAM_GRID_RFC;

In [36]:
%%time
model_RFC, F1_RFC = best_model(features_train, 
                  target_train,
                  RandomForestClassifier(),
                  PARAM_GRID_RFC, 
                  5, 
                  'F1')

{'max_depth': 4, 'min_samples_leaf': 3, 'n_estimators': 3, 'random_state': 12345}
F1 : 0.18489049978423155
Wall time: 3min 30s


{'max_depth': 4, 'min_samples_leaf': 3, 'n_estimators': 3, 'random_state': 12345}
F1 : 0.18489049978423155
Wall time: 3min 30s

In [37]:
# Сетка параметров для HistGradientBoostingClassifier()
PARAM_GRID_HGBC = {'learning_rate': [0.05, 0.1, 0.2, 0.5, 0.6, 0.7, 0.8, 0.9], #[0.2 ,0.5, 0.7]
                   #'loss': ['auto', 'binary_crossentropy', 'categorical_crossentropy']
                  'max_iter': [   1,   10,  100, 200, 300, 400], #np.logspace(0, 2, 5, endpoint=True, dtype = 'int'),
                  #'max_depth': range(1,5),
                  'random_state': [12345],
                  #'verbose': [1]
                  }  
PARAM_GRID_HGBC;

In [38]:
%%time
model_HGBC, F1_HGBC = best_model(features_train, 
                  target_train,
                  HistGradientBoostingClassifier(),
                  PARAM_GRID_HGBC, 
                  5, 
                  'F1')

{'learning_rate': 0.5, 'max_iter': 300, 'random_state': 12345}
F1 : 0.5590105930737465
Wall time: 48min 46s


{'learning_rate': 0.5, 'max_iter': 300, 'random_state': 12345}
F1 : 0.5590105930737465
Wall time: 48min 46s

In [55]:
# Сетка параметров для CatBoostClassifier()
PARAM_GRID_CBC = {'learning_rate': [0.05, 0.2]#, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9], 
                  'iterations': [100, 200, 300]#   10,  100, 200], #np.logspace(0, 2, 5, endpoint=True, dtype = 'int'),
                  #'verbose': [10],
                  }  
PARAM_GRID_CBC;

In [56]:
%%time
model_CBC, F1_CBC = best_model(features_train, 
                  target_train,
                  CatBoostClassifier(),
                  PARAM_GRID_CBC, 
                  5, 
                  'F1')

0:	learn: 0.5183328	total: 95.2ms	remaining: 9.43s
1:	learn: 0.4087278	total: 189ms	remaining: 9.26s
2:	learn: 0.3337986	total: 284ms	remaining: 9.19s
3:	learn: 0.2887790	total: 386ms	remaining: 9.26s
4:	learn: 0.2532627	total: 481ms	remaining: 9.13s
5:	learn: 0.2285494	total: 579ms	remaining: 9.07s
6:	learn: 0.2141317	total: 668ms	remaining: 8.87s
7:	learn: 0.2026738	total: 764ms	remaining: 8.78s
8:	learn: 0.1898714	total: 854ms	remaining: 8.64s
9:	learn: 0.1807304	total: 954ms	remaining: 8.59s
10:	learn: 0.1731961	total: 1.05s	remaining: 8.47s
11:	learn: 0.1660728	total: 1.15s	remaining: 8.43s
12:	learn: 0.1603077	total: 1.25s	remaining: 8.34s
13:	learn: 0.1551958	total: 1.35s	remaining: 8.28s
14:	learn: 0.1507237	total: 1.44s	remaining: 8.18s
15:	learn: 0.1457535	total: 1.54s	remaining: 8.09s
16:	learn: 0.1418714	total: 1.63s	remaining: 7.96s
17:	learn: 0.1370607	total: 1.73s	remaining: 7.86s
18:	learn: 0.1333593	total: 1.81s	remaining: 7.74s
19:	learn: 0.1299575	total: 1.91s	remain

62:	learn: 0.0457377	total: 6.18s	remaining: 3.63s
63:	learn: 0.0451512	total: 6.28s	remaining: 3.53s
64:	learn: 0.0442104	total: 6.38s	remaining: 3.44s
65:	learn: 0.0433697	total: 6.48s	remaining: 3.34s
66:	learn: 0.0429784	total: 6.57s	remaining: 3.24s
67:	learn: 0.0420000	total: 6.66s	remaining: 3.13s
68:	learn: 0.0409819	total: 6.76s	remaining: 3.04s
69:	learn: 0.0401225	total: 6.85s	remaining: 2.94s
70:	learn: 0.0394253	total: 6.95s	remaining: 2.84s
71:	learn: 0.0385227	total: 7.04s	remaining: 2.74s
72:	learn: 0.0376331	total: 7.14s	remaining: 2.64s
73:	learn: 0.0368706	total: 7.24s	remaining: 2.54s
74:	learn: 0.0362912	total: 7.34s	remaining: 2.45s
75:	learn: 0.0353170	total: 7.44s	remaining: 2.35s
76:	learn: 0.0348687	total: 7.54s	remaining: 2.25s
77:	learn: 0.0346222	total: 7.63s	remaining: 2.15s
78:	learn: 0.0336221	total: 7.72s	remaining: 2.05s
79:	learn: 0.0329587	total: 7.82s	remaining: 1.96s
80:	learn: 0.0324096	total: 7.92s	remaining: 1.86s
81:	learn: 0.0316506	total: 8.0

24:	learn: 0.1080674	total: 2.75s	remaining: 8.24s
25:	learn: 0.1049514	total: 2.85s	remaining: 8.13s
26:	learn: 0.1032068	total: 2.96s	remaining: 7.99s
27:	learn: 0.1007038	total: 3.06s	remaining: 7.88s
28:	learn: 0.0985040	total: 3.18s	remaining: 7.79s
29:	learn: 0.0956752	total: 3.31s	remaining: 7.72s
30:	learn: 0.0935694	total: 3.43s	remaining: 7.63s
31:	learn: 0.0919623	total: 3.54s	remaining: 7.53s
32:	learn: 0.0905950	total: 3.66s	remaining: 7.43s
33:	learn: 0.0882465	total: 3.77s	remaining: 7.31s
34:	learn: 0.0855737	total: 3.87s	remaining: 7.18s
35:	learn: 0.0824908	total: 3.97s	remaining: 7.06s
36:	learn: 0.0812682	total: 4.07s	remaining: 6.93s
37:	learn: 0.0794211	total: 4.18s	remaining: 6.82s
38:	learn: 0.0771659	total: 4.29s	remaining: 6.72s
39:	learn: 0.0757881	total: 4.4s	remaining: 6.61s
40:	learn: 0.0740499	total: 4.51s	remaining: 6.49s
41:	learn: 0.0722385	total: 4.62s	remaining: 6.38s
42:	learn: 0.0710667	total: 4.73s	remaining: 6.26s
43:	learn: 0.0699612	total: 4.83

86:	learn: 0.0284647	total: 9.42s	remaining: 1.41s
87:	learn: 0.0275672	total: 9.52s	remaining: 1.3s
88:	learn: 0.0272716	total: 9.62s	remaining: 1.19s
89:	learn: 0.0268361	total: 9.73s	remaining: 1.08s
90:	learn: 0.0264495	total: 9.83s	remaining: 972ms
91:	learn: 0.0261358	total: 9.94s	remaining: 864ms
92:	learn: 0.0256877	total: 10s	remaining: 756ms
93:	learn: 0.0253200	total: 10.2s	remaining: 648ms
94:	learn: 0.0250344	total: 10.3s	remaining: 540ms
95:	learn: 0.0243081	total: 10.4s	remaining: 432ms
96:	learn: 0.0240508	total: 10.5s	remaining: 324ms
97:	learn: 0.0234289	total: 10.6s	remaining: 216ms
98:	learn: 0.0231890	total: 10.7s	remaining: 108ms
99:	learn: 0.0223850	total: 10.8s	remaining: 0us
0:	learn: 0.4535525	total: 112ms	remaining: 11.1s
1:	learn: 0.3418698	total: 212ms	remaining: 10.4s
2:	learn: 0.2769111	total: 323ms	remaining: 10.4s
3:	learn: 0.2375101	total: 424ms	remaining: 10.2s
4:	learn: 0.2136130	total: 523ms	remaining: 9.93s
5:	learn: 0.1962600	total: 627ms	remainin

48:	learn: 0.0369613	total: 5.3s	remaining: 5.51s
49:	learn: 0.0361976	total: 5.41s	remaining: 5.41s
50:	learn: 0.0354690	total: 5.51s	remaining: 5.29s
51:	learn: 0.0343027	total: 5.61s	remaining: 5.18s
52:	learn: 0.0332675	total: 5.71s	remaining: 5.07s
53:	learn: 0.0323437	total: 5.82s	remaining: 4.96s
54:	learn: 0.0309793	total: 5.93s	remaining: 4.85s
55:	learn: 0.0297881	total: 6.04s	remaining: 4.75s
56:	learn: 0.0289248	total: 6.16s	remaining: 4.64s
57:	learn: 0.0284061	total: 6.26s	remaining: 4.53s
58:	learn: 0.0275407	total: 6.37s	remaining: 4.43s
59:	learn: 0.0267890	total: 6.47s	remaining: 4.32s
60:	learn: 0.0256850	total: 6.58s	remaining: 4.2s
61:	learn: 0.0248336	total: 6.68s	remaining: 4.1s
62:	learn: 0.0242293	total: 6.8s	remaining: 3.99s
63:	learn: 0.0237434	total: 6.91s	remaining: 3.88s
64:	learn: 0.0230925	total: 7.02s	remaining: 3.78s
65:	learn: 0.0226994	total: 7.12s	remaining: 3.67s
66:	learn: 0.0220784	total: 7.23s	remaining: 3.56s
67:	learn: 0.0217693	total: 7.33s	r

10:	learn: 0.1424970	total: 1.17s	remaining: 9.46s
11:	learn: 0.1380287	total: 1.28s	remaining: 9.41s
12:	learn: 0.1309295	total: 1.39s	remaining: 9.34s
13:	learn: 0.1244111	total: 1.5s	remaining: 9.24s
14:	learn: 0.1197696	total: 1.61s	remaining: 9.13s
15:	learn: 0.1158471	total: 1.72s	remaining: 9.02s
16:	learn: 0.1106433	total: 1.82s	remaining: 8.88s
17:	learn: 0.1077533	total: 1.92s	remaining: 8.75s
18:	learn: 0.1034896	total: 2.02s	remaining: 8.63s
19:	learn: 0.1003153	total: 2.12s	remaining: 8.5s
20:	learn: 0.0969834	total: 2.23s	remaining: 8.38s
21:	learn: 0.0934058	total: 2.33s	remaining: 8.28s
22:	learn: 0.0895046	total: 2.45s	remaining: 8.19s
23:	learn: 0.0860452	total: 2.56s	remaining: 8.1s
24:	learn: 0.0832098	total: 2.66s	remaining: 7.99s
25:	learn: 0.0800975	total: 2.77s	remaining: 7.88s
26:	learn: 0.0780480	total: 2.87s	remaining: 7.75s
27:	learn: 0.0754066	total: 2.97s	remaining: 7.64s
28:	learn: 0.0723012	total: 3.07s	remaining: 7.53s
29:	learn: 0.0694643	total: 3.18s	

72:	learn: 0.0219135	total: 8.17s	remaining: 3.02s
73:	learn: 0.0213814	total: 8.29s	remaining: 2.91s
74:	learn: 0.0206288	total: 8.41s	remaining: 2.8s
75:	learn: 0.0200453	total: 8.52s	remaining: 2.69s
76:	learn: 0.0195434	total: 8.65s	remaining: 2.58s
77:	learn: 0.0193183	total: 8.76s	remaining: 2.47s
78:	learn: 0.0190178	total: 8.87s	remaining: 2.36s
79:	learn: 0.0183471	total: 8.99s	remaining: 2.25s
80:	learn: 0.0180420	total: 9.12s	remaining: 2.14s
81:	learn: 0.0175999	total: 9.27s	remaining: 2.03s
82:	learn: 0.0172879	total: 9.42s	remaining: 1.93s
83:	learn: 0.0170850	total: 9.54s	remaining: 1.82s
84:	learn: 0.0166273	total: 9.68s	remaining: 1.71s
85:	learn: 0.0164225	total: 9.8s	remaining: 1.59s
86:	learn: 0.0158612	total: 9.94s	remaining: 1.49s
87:	learn: 0.0155375	total: 10.1s	remaining: 1.37s
88:	learn: 0.0153673	total: 10.2s	remaining: 1.26s
89:	learn: 0.0150439	total: 10.3s	remaining: 1.15s
90:	learn: 0.0147615	total: 10.4s	remaining: 1.03s
91:	learn: 0.0143680	total: 10.6s

34:	learn: 0.0408730	total: 3.88s	remaining: 7.21s
35:	learn: 0.0382619	total: 4.01s	remaining: 7.13s
36:	learn: 0.0371033	total: 4.12s	remaining: 7.02s
37:	learn: 0.0364433	total: 4.24s	remaining: 6.91s
38:	learn: 0.0353225	total: 4.35s	remaining: 6.81s
39:	learn: 0.0345012	total: 4.46s	remaining: 6.69s
40:	learn: 0.0333623	total: 4.57s	remaining: 6.58s
41:	learn: 0.0321375	total: 4.68s	remaining: 6.46s
42:	learn: 0.0312998	total: 4.79s	remaining: 6.35s
43:	learn: 0.0308892	total: 4.89s	remaining: 6.22s
44:	learn: 0.0304300	total: 5s	remaining: 6.11s
45:	learn: 0.0296712	total: 5.11s	remaining: 6s
46:	learn: 0.0279163	total: 5.22s	remaining: 5.89s
47:	learn: 0.0268712	total: 5.33s	remaining: 5.78s
48:	learn: 0.0262116	total: 5.45s	remaining: 5.67s
49:	learn: 0.0258497	total: 5.56s	remaining: 5.56s
50:	learn: 0.0254867	total: 5.66s	remaining: 5.44s
51:	learn: 0.0244739	total: 5.77s	remaining: 5.32s
52:	learn: 0.0236521	total: 5.87s	remaining: 5.21s
53:	learn: 0.0230455	total: 5.98s	rem

97:	learn: 0.0082901	total: 10.3s	remaining: 211ms
98:	learn: 0.0079899	total: 10.4s	remaining: 105ms
99:	learn: 0.0077611	total: 10.5s	remaining: 0us
0:	learn: 0.3779609	total: 156ms	remaining: 15.5s
1:	learn: 0.2741967	total: 257ms	remaining: 12.6s
2:	learn: 0.2234116	total: 361ms	remaining: 11.7s
3:	learn: 0.1967858	total: 467ms	remaining: 11.2s
4:	learn: 0.1782280	total: 569ms	remaining: 10.8s
5:	learn: 0.1605004	total: 667ms	remaining: 10.5s
6:	learn: 0.1529418	total: 768ms	remaining: 10.2s
7:	learn: 0.1449621	total: 868ms	remaining: 9.98s
8:	learn: 0.1378752	total: 969ms	remaining: 9.8s
9:	learn: 0.1300137	total: 1.07s	remaining: 9.66s
10:	learn: 0.1247256	total: 1.18s	remaining: 9.57s
11:	learn: 0.1203587	total: 1.29s	remaining: 9.44s
12:	learn: 0.1140281	total: 1.39s	remaining: 9.33s
13:	learn: 0.1093630	total: 1.5s	remaining: 9.24s
14:	learn: 0.1025482	total: 1.61s	remaining: 9.1s
15:	learn: 0.0987818	total: 1.71s	remaining: 8.95s
16:	learn: 0.0935903	total: 1.81s	remaining: 8

59:	learn: 0.0205733	total: 6.28s	remaining: 4.18s
60:	learn: 0.0191426	total: 6.38s	remaining: 4.08s
61:	learn: 0.0188274	total: 6.49s	remaining: 3.98s
62:	learn: 0.0187400	total: 6.6s	remaining: 3.88s
63:	learn: 0.0182822	total: 6.71s	remaining: 3.77s
64:	learn: 0.0176697	total: 6.81s	remaining: 3.67s
65:	learn: 0.0171817	total: 6.91s	remaining: 3.56s
66:	learn: 0.0170956	total: 7.01s	remaining: 3.45s
67:	learn: 0.0169326	total: 7.11s	remaining: 3.34s
68:	learn: 0.0164859	total: 7.2s	remaining: 3.24s
69:	learn: 0.0158872	total: 7.31s	remaining: 3.13s
70:	learn: 0.0152472	total: 7.42s	remaining: 3.03s
71:	learn: 0.0149179	total: 7.53s	remaining: 2.93s
72:	learn: 0.0143192	total: 7.63s	remaining: 2.82s
73:	learn: 0.0141357	total: 7.74s	remaining: 2.72s
74:	learn: 0.0136828	total: 7.84s	remaining: 2.61s
75:	learn: 0.0133591	total: 7.94s	remaining: 2.51s
76:	learn: 0.0129305	total: 8.04s	remaining: 2.4s
77:	learn: 0.0127194	total: 8.14s	remaining: 2.3s
78:	learn: 0.0126190	total: 8.24s	r

22:	learn: 0.0601411	total: 2.41s	remaining: 8.06s
23:	learn: 0.0554121	total: 2.51s	remaining: 7.95s
24:	learn: 0.0537589	total: 2.61s	remaining: 7.83s
25:	learn: 0.0487970	total: 2.71s	remaining: 7.72s
26:	learn: 0.0475267	total: 2.81s	remaining: 7.6s
27:	learn: 0.0467913	total: 2.92s	remaining: 7.51s
28:	learn: 0.0432754	total: 3.03s	remaining: 7.42s
29:	learn: 0.0402390	total: 3.13s	remaining: 7.32s
30:	learn: 0.0381042	total: 3.24s	remaining: 7.22s
31:	learn: 0.0365433	total: 3.35s	remaining: 7.13s
32:	learn: 0.0330545	total: 3.45s	remaining: 7.01s
33:	learn: 0.0323968	total: 3.55s	remaining: 6.89s
34:	learn: 0.0304271	total: 3.65s	remaining: 6.78s
35:	learn: 0.0290155	total: 3.75s	remaining: 6.67s
36:	learn: 0.0278724	total: 3.85s	remaining: 6.56s
37:	learn: 0.0269765	total: 3.96s	remaining: 6.46s
38:	learn: 0.0261324	total: 4.07s	remaining: 6.36s
39:	learn: 0.0247026	total: 4.18s	remaining: 6.27s
40:	learn: 0.0240698	total: 4.29s	remaining: 6.17s
41:	learn: 0.0234359	total: 4.39

86:	learn: 0.0067234	total: 9.04s	remaining: 1.35s
87:	learn: 0.0064913	total: 9.15s	remaining: 1.25s
88:	learn: 0.0063700	total: 9.25s	remaining: 1.14s
89:	learn: 0.0062689	total: 9.36s	remaining: 1.04s
90:	learn: 0.0061142	total: 9.47s	remaining: 936ms
91:	learn: 0.0059004	total: 9.57s	remaining: 833ms
92:	learn: 0.0058122	total: 9.67s	remaining: 728ms
93:	learn: 0.0056731	total: 9.77s	remaining: 624ms
94:	learn: 0.0055489	total: 9.87s	remaining: 520ms
95:	learn: 0.0053942	total: 9.97s	remaining: 416ms
96:	learn: 0.0052447	total: 10.1s	remaining: 312ms
97:	learn: 0.0051157	total: 10.2s	remaining: 208ms
98:	learn: 0.0049989	total: 10.3s	remaining: 104ms
99:	learn: 0.0049490	total: 10.4s	remaining: 0us
0:	learn: 0.3381769	total: 105ms	remaining: 10.4s
1:	learn: 0.2416487	total: 203ms	remaining: 9.94s
2:	learn: 0.2067097	total: 303ms	remaining: 9.8s
3:	learn: 0.1853000	total: 407ms	remaining: 9.76s
4:	learn: 0.1720094	total: 510ms	remaining: 9.69s
5:	learn: 0.1595090	total: 617ms	remain

48:	learn: 0.0203092	total: 5.17s	remaining: 5.38s
49:	learn: 0.0193007	total: 5.28s	remaining: 5.28s
50:	learn: 0.0187211	total: 5.38s	remaining: 5.17s
51:	learn: 0.0180978	total: 5.49s	remaining: 5.06s
52:	learn: 0.0175678	total: 5.59s	remaining: 4.96s
53:	learn: 0.0169149	total: 5.69s	remaining: 4.85s
54:	learn: 0.0162731	total: 5.79s	remaining: 4.74s
55:	learn: 0.0156366	total: 5.91s	remaining: 4.64s
56:	learn: 0.0151801	total: 6.02s	remaining: 4.54s
57:	learn: 0.0150281	total: 6.13s	remaining: 4.44s
58:	learn: 0.0146129	total: 6.24s	remaining: 4.33s
59:	learn: 0.0141929	total: 6.34s	remaining: 4.23s
60:	learn: 0.0136874	total: 6.44s	remaining: 4.12s
61:	learn: 0.0131806	total: 6.55s	remaining: 4.01s
62:	learn: 0.0130151	total: 6.65s	remaining: 3.9s
63:	learn: 0.0127663	total: 6.75s	remaining: 3.8s
64:	learn: 0.0121383	total: 6.86s	remaining: 3.69s
65:	learn: 0.0117731	total: 6.97s	remaining: 3.59s
66:	learn: 0.0114928	total: 7.08s	remaining: 3.49s
67:	learn: 0.0112299	total: 7.19s

11:	learn: 0.0896135	total: 1.29s	remaining: 9.49s
12:	learn: 0.0861721	total: 1.4s	remaining: 9.38s
13:	learn: 0.0808603	total: 1.51s	remaining: 9.26s
14:	learn: 0.0752465	total: 1.61s	remaining: 9.14s
15:	learn: 0.0721849	total: 1.72s	remaining: 9.03s
16:	learn: 0.0674687	total: 1.82s	remaining: 8.9s
17:	learn: 0.0653631	total: 1.92s	remaining: 8.76s
18:	learn: 0.0621147	total: 2.03s	remaining: 8.64s
19:	learn: 0.0601971	total: 2.14s	remaining: 8.56s
20:	learn: 0.0557503	total: 2.25s	remaining: 8.47s
21:	learn: 0.0512042	total: 2.36s	remaining: 8.37s
22:	learn: 0.0477944	total: 2.47s	remaining: 8.27s
23:	learn: 0.0460912	total: 2.57s	remaining: 8.14s
24:	learn: 0.0437029	total: 2.67s	remaining: 8.01s
25:	learn: 0.0412313	total: 2.77s	remaining: 7.9s
26:	learn: 0.0394219	total: 2.88s	remaining: 7.79s
27:	learn: 0.0376530	total: 2.98s	remaining: 7.67s
28:	learn: 0.0345697	total: 3.1s	remaining: 7.58s
29:	learn: 0.0340565	total: 3.21s	remaining: 7.48s
30:	learn: 0.0315982	total: 3.31s	r

73:	learn: 0.0063839	total: 7.8s	remaining: 2.74s
74:	learn: 0.0063142	total: 7.91s	remaining: 2.63s
75:	learn: 0.0060571	total: 8.01s	remaining: 2.53s
76:	learn: 0.0059292	total: 8.12s	remaining: 2.42s
77:	learn: 0.0056516	total: 8.22s	remaining: 2.32s
78:	learn: 0.0055044	total: 8.34s	remaining: 2.21s
79:	learn: 0.0052630	total: 8.44s	remaining: 2.11s
80:	learn: 0.0052020	total: 8.55s	remaining: 2s
81:	learn: 0.0051847	total: 8.65s	remaining: 1.9s
82:	learn: 0.0050503	total: 8.75s	remaining: 1.79s
83:	learn: 0.0048851	total: 8.85s	remaining: 1.69s
84:	learn: 0.0046176	total: 8.96s	remaining: 1.58s
85:	learn: 0.0044729	total: 9.06s	remaining: 1.48s
86:	learn: 0.0044034	total: 9.17s	remaining: 1.37s
87:	learn: 0.0043441	total: 9.28s	remaining: 1.26s
88:	learn: 0.0042635	total: 9.39s	remaining: 1.16s
89:	learn: 0.0042153	total: 9.49s	remaining: 1.05s
90:	learn: 0.0041594	total: 9.6s	remaining: 949ms
91:	learn: 0.0040480	total: 9.7s	remaining: 844ms
92:	learn: 0.0039141	total: 9.8s	remai

36:	learn: 0.0250462	total: 3.91s	remaining: 6.65s
37:	learn: 0.0243703	total: 4.01s	remaining: 6.54s
38:	learn: 0.0239326	total: 4.11s	remaining: 6.42s
39:	learn: 0.0226193	total: 4.21s	remaining: 6.31s
40:	learn: 0.0218786	total: 4.31s	remaining: 6.2s
41:	learn: 0.0217022	total: 4.41s	remaining: 6.09s
42:	learn: 0.0206904	total: 4.52s	remaining: 5.99s
43:	learn: 0.0202141	total: 4.63s	remaining: 5.89s
44:	learn: 0.0193757	total: 4.74s	remaining: 5.79s
45:	learn: 0.0188873	total: 4.86s	remaining: 5.7s
46:	learn: 0.0180514	total: 4.96s	remaining: 5.59s
47:	learn: 0.0173490	total: 5.06s	remaining: 5.48s
48:	learn: 0.0164707	total: 5.16s	remaining: 5.37s
49:	learn: 0.0156175	total: 5.26s	remaining: 5.26s
50:	learn: 0.0150069	total: 5.36s	remaining: 5.15s
51:	learn: 0.0137979	total: 5.47s	remaining: 5.05s
52:	learn: 0.0136263	total: 5.58s	remaining: 4.95s
53:	learn: 0.0132424	total: 5.68s	remaining: 4.84s
54:	learn: 0.0126203	total: 5.79s	remaining: 4.74s
55:	learn: 0.0120682	total: 5.9s	

98:	learn: 0.0031343	total: 10.6s	remaining: 107ms
99:	learn: 0.0030372	total: 10.7s	remaining: 0us
0:	learn: 0.2806108	total: 106ms	remaining: 10.5s
1:	learn: 0.2186354	total: 204ms	remaining: 10s
2:	learn: 0.1912892	total: 305ms	remaining: 9.88s
3:	learn: 0.1720017	total: 406ms	remaining: 9.73s
4:	learn: 0.1586028	total: 506ms	remaining: 9.62s
5:	learn: 0.1503925	total: 609ms	remaining: 9.54s
6:	learn: 0.1363241	total: 715ms	remaining: 9.5s
7:	learn: 0.1238689	total: 822ms	remaining: 9.45s
8:	learn: 0.1178336	total: 930ms	remaining: 9.4s
9:	learn: 0.1082334	total: 1.04s	remaining: 9.35s
10:	learn: 0.0984755	total: 1.14s	remaining: 9.25s
11:	learn: 0.0934946	total: 1.25s	remaining: 9.15s
12:	learn: 0.0853392	total: 1.36s	remaining: 9.09s
13:	learn: 0.0787386	total: 1.46s	remaining: 8.98s
14:	learn: 0.0739770	total: 1.57s	remaining: 8.88s
15:	learn: 0.0654944	total: 1.67s	remaining: 8.79s
16:	learn: 0.0610182	total: 1.78s	remaining: 8.71s
17:	learn: 0.0545317	total: 1.89s	remaining: 8.

61:	learn: 0.0084004	total: 6.47s	remaining: 3.96s
62:	learn: 0.0080901	total: 6.57s	remaining: 3.86s
63:	learn: 0.0076009	total: 6.67s	remaining: 3.75s
64:	learn: 0.0073412	total: 6.78s	remaining: 3.65s
65:	learn: 0.0071620	total: 6.88s	remaining: 3.54s
66:	learn: 0.0067825	total: 6.99s	remaining: 3.44s
67:	learn: 0.0067108	total: 7.09s	remaining: 3.34s
68:	learn: 0.0066793	total: 7.2s	remaining: 3.23s
69:	learn: 0.0065948	total: 7.3s	remaining: 3.13s
70:	learn: 0.0064772	total: 7.41s	remaining: 3.02s
71:	learn: 0.0062783	total: 7.51s	remaining: 2.92s
72:	learn: 0.0062192	total: 7.61s	remaining: 2.81s
73:	learn: 0.0060806	total: 7.71s	remaining: 2.71s
74:	learn: 0.0058413	total: 7.81s	remaining: 2.6s
75:	learn: 0.0055889	total: 7.91s	remaining: 2.5s
76:	learn: 0.0055283	total: 8.02s	remaining: 2.4s
77:	learn: 0.0053783	total: 8.13s	remaining: 2.29s
78:	learn: 0.0052329	total: 8.23s	remaining: 2.19s
79:	learn: 0.0051800	total: 8.34s	remaining: 2.08s
80:	learn: 0.0050210	total: 8.44s	re

23:	learn: 0.0403408	total: 2.47s	remaining: 7.83s
24:	learn: 0.0376460	total: 2.58s	remaining: 7.74s
25:	learn: 0.0341412	total: 2.69s	remaining: 7.65s
26:	learn: 0.0330216	total: 2.8s	remaining: 7.56s
27:	learn: 0.0323646	total: 2.9s	remaining: 7.46s
28:	learn: 0.0311427	total: 3.01s	remaining: 7.36s
29:	learn: 0.0283965	total: 3.11s	remaining: 7.26s
30:	learn: 0.0268834	total: 3.21s	remaining: 7.14s
31:	learn: 0.0259542	total: 3.31s	remaining: 7.03s
32:	learn: 0.0245092	total: 3.41s	remaining: 6.92s
33:	learn: 0.0240970	total: 3.51s	remaining: 6.81s
34:	learn: 0.0229557	total: 3.62s	remaining: 6.72s
35:	learn: 0.0220291	total: 3.73s	remaining: 6.63s
36:	learn: 0.0211700	total: 3.83s	remaining: 6.53s
37:	learn: 0.0196568	total: 3.94s	remaining: 6.43s
38:	learn: 0.0192210	total: 4.04s	remaining: 6.32s
39:	learn: 0.0181951	total: 4.14s	remaining: 6.21s
40:	learn: 0.0165121	total: 4.24s	remaining: 6.11s
41:	learn: 0.0153956	total: 4.34s	remaining: 6s
42:	learn: 0.0143729	total: 4.45s	re

87:	learn: 0.0032275	total: 9.15s	remaining: 1.25s
88:	learn: 0.0031273	total: 9.26s	remaining: 1.14s
89:	learn: 0.0030583	total: 9.37s	remaining: 1.04s
90:	learn: 0.0030149	total: 9.47s	remaining: 937ms
91:	learn: 0.0029836	total: 9.57s	remaining: 832ms
92:	learn: 0.0028355	total: 9.67s	remaining: 728ms
93:	learn: 0.0027614	total: 9.78s	remaining: 624ms
94:	learn: 0.0027422	total: 9.89s	remaining: 520ms
95:	learn: 0.0027048	total: 9.99s	remaining: 416ms
96:	learn: 0.0026622	total: 10.1s	remaining: 312ms
97:	learn: 0.0026302	total: 10.2s	remaining: 208ms
98:	learn: 0.0026205	total: 10.3s	remaining: 104ms
99:	learn: 0.0025318	total: 10.4s	remaining: 0us
0:	learn: 0.2678176	total: 152ms	remaining: 15s
1:	learn: 0.2064802	total: 250ms	remaining: 12.3s
2:	learn: 0.1797767	total: 355ms	remaining: 11.5s
3:	learn: 0.1610072	total: 482ms	remaining: 11.6s
4:	learn: 0.1452022	total: 644ms	remaining: 12.2s
5:	learn: 0.1331234	total: 752ms	remaining: 11.8s
6:	learn: 0.1184014	total: 865ms	remainin

49:	learn: 0.0098494	total: 5.41s	remaining: 5.41s
50:	learn: 0.0097506	total: 5.51s	remaining: 5.3s
51:	learn: 0.0095592	total: 5.63s	remaining: 5.19s
52:	learn: 0.0092032	total: 5.74s	remaining: 5.09s
53:	learn: 0.0087021	total: 5.85s	remaining: 4.98s
54:	learn: 0.0080550	total: 5.95s	remaining: 4.87s
55:	learn: 0.0078997	total: 6.06s	remaining: 4.76s
56:	learn: 0.0076729	total: 6.17s	remaining: 4.65s
57:	learn: 0.0074917	total: 6.27s	remaining: 4.54s
58:	learn: 0.0072616	total: 6.37s	remaining: 4.43s
59:	learn: 0.0069662	total: 6.48s	remaining: 4.32s
60:	learn: 0.0067392	total: 6.59s	remaining: 4.21s
61:	learn: 0.0064210	total: 6.71s	remaining: 4.11s
62:	learn: 0.0061051	total: 6.83s	remaining: 4.01s
63:	learn: 0.0059740	total: 6.94s	remaining: 3.9s
64:	learn: 0.0057847	total: 7.04s	remaining: 3.79s
65:	learn: 0.0054624	total: 7.14s	remaining: 3.68s
66:	learn: 0.0052178	total: 7.24s	remaining: 3.57s
67:	learn: 0.0050694	total: 7.35s	remaining: 3.46s
68:	learn: 0.0049091	total: 7.45s

12:	learn: 0.0738111	total: 1.39s	remaining: 9.32s
13:	learn: 0.0696577	total: 1.51s	remaining: 9.26s
14:	learn: 0.0665593	total: 1.63s	remaining: 9.21s
15:	learn: 0.0618550	total: 1.74s	remaining: 9.11s
16:	learn: 0.0575789	total: 1.84s	remaining: 8.98s
17:	learn: 0.0559245	total: 1.94s	remaining: 8.84s
18:	learn: 0.0498238	total: 2.04s	remaining: 8.7s
19:	learn: 0.0484313	total: 2.14s	remaining: 8.57s
20:	learn: 0.0449343	total: 2.25s	remaining: 8.45s
21:	learn: 0.0431911	total: 2.36s	remaining: 8.36s
22:	learn: 0.0409088	total: 2.47s	remaining: 8.29s
23:	learn: 0.0385741	total: 2.59s	remaining: 8.21s
24:	learn: 0.0344159	total: 2.7s	remaining: 8.11s
25:	learn: 0.0327650	total: 2.81s	remaining: 8s
26:	learn: 0.0307229	total: 2.92s	remaining: 7.9s
27:	learn: 0.0295542	total: 3.02s	remaining: 7.77s
28:	learn: 0.0279655	total: 3.12s	remaining: 7.64s
29:	learn: 0.0266872	total: 3.23s	remaining: 7.54s
30:	learn: 0.0256186	total: 3.35s	remaining: 7.45s
31:	learn: 0.0236623	total: 3.46s	rem

75:	learn: 0.0034466	total: 8.36s	remaining: 2.64s
76:	learn: 0.0032890	total: 8.47s	remaining: 2.53s
77:	learn: 0.0032566	total: 8.57s	remaining: 2.42s
78:	learn: 0.0031870	total: 8.67s	remaining: 2.31s
79:	learn: 0.0030733	total: 8.78s	remaining: 2.19s
80:	learn: 0.0030023	total: 8.88s	remaining: 2.08s
81:	learn: 0.0029491	total: 8.98s	remaining: 1.97s
82:	learn: 0.0028258	total: 9.09s	remaining: 1.86s
83:	learn: 0.0027501	total: 9.2s	remaining: 1.75s
84:	learn: 0.0027226	total: 9.31s	remaining: 1.64s
85:	learn: 0.0027166	total: 9.42s	remaining: 1.53s
86:	learn: 0.0026962	total: 9.54s	remaining: 1.43s
87:	learn: 0.0026348	total: 9.64s	remaining: 1.31s
88:	learn: 0.0025560	total: 9.74s	remaining: 1.2s
89:	learn: 0.0025289	total: 9.85s	remaining: 1.09s
90:	learn: 0.0025078	total: 9.95s	remaining: 984ms
91:	learn: 0.0024809	total: 10.1s	remaining: 874ms
92:	learn: 0.0024373	total: 10.2s	remaining: 765ms
93:	learn: 0.0023768	total: 10.3s	remaining: 656ms
94:	learn: 0.0023193	total: 10.4s

38:	learn: 0.0147060	total: 4.27s	remaining: 6.68s
39:	learn: 0.0142340	total: 4.38s	remaining: 6.57s
40:	learn: 0.0136111	total: 4.48s	remaining: 6.44s
41:	learn: 0.0123137	total: 4.58s	remaining: 6.33s
42:	learn: 0.0118306	total: 4.69s	remaining: 6.22s
43:	learn: 0.0111745	total: 4.81s	remaining: 6.12s
44:	learn: 0.0106061	total: 4.93s	remaining: 6.02s
45:	learn: 0.0101734	total: 5.04s	remaining: 5.91s
46:	learn: 0.0098863	total: 5.15s	remaining: 5.8s
47:	learn: 0.0092790	total: 5.27s	remaining: 5.71s
48:	learn: 0.0091474	total: 5.39s	remaining: 5.61s
49:	learn: 0.0084883	total: 5.49s	remaining: 5.49s
50:	learn: 0.0083106	total: 5.59s	remaining: 5.37s
51:	learn: 0.0078644	total: 5.7s	remaining: 5.26s
52:	learn: 0.0074603	total: 5.8s	remaining: 5.14s
53:	learn: 0.0073604	total: 5.91s	remaining: 5.03s
54:	learn: 0.0072204	total: 6.02s	remaining: 4.93s
55:	learn: 0.0067561	total: 6.13s	remaining: 4.82s
56:	learn: 0.0064654	total: 6.25s	remaining: 4.71s
57:	learn: 0.0063122	total: 6.35s	

0:	learn: 0.2829640	total: 133ms	remaining: 13.1s
1:	learn: 0.2061716	total: 234ms	remaining: 11.5s
2:	learn: 0.1769497	total: 340ms	remaining: 11s
3:	learn: 0.1540541	total: 447ms	remaining: 10.7s
4:	learn: 0.1409935	total: 548ms	remaining: 10.4s
5:	learn: 0.1271489	total: 651ms	remaining: 10.2s
6:	learn: 0.1141577	total: 751ms	remaining: 9.97s
7:	learn: 0.1058603	total: 859ms	remaining: 9.87s
8:	learn: 0.0969741	total: 967ms	remaining: 9.78s
9:	learn: 0.0894982	total: 1.07s	remaining: 9.67s
10:	learn: 0.0844533	total: 1.18s	remaining: 9.57s
11:	learn: 0.0775351	total: 1.29s	remaining: 9.48s
12:	learn: 0.0679846	total: 1.41s	remaining: 9.41s
13:	learn: 0.0611095	total: 1.51s	remaining: 9.29s
14:	learn: 0.0550419	total: 1.62s	remaining: 9.19s
15:	learn: 0.0505580	total: 1.73s	remaining: 9.07s
16:	learn: 0.0471836	total: 1.84s	remaining: 9s
17:	learn: 0.0418473	total: 1.96s	remaining: 8.92s
18:	learn: 0.0394604	total: 2.07s	remaining: 8.84s
19:	learn: 0.0368128	total: 2.19s	remaining: 8

62:	learn: 0.0527713	total: 6.91s	remaining: 15s
63:	learn: 0.0511797	total: 7.02s	remaining: 14.9s
64:	learn: 0.0501285	total: 7.14s	remaining: 14.8s
65:	learn: 0.0492289	total: 7.25s	remaining: 14.7s
66:	learn: 0.0482691	total: 7.35s	remaining: 14.6s
67:	learn: 0.0474497	total: 7.46s	remaining: 14.5s
68:	learn: 0.0466356	total: 7.56s	remaining: 14.3s
69:	learn: 0.0455390	total: 7.67s	remaining: 14.2s
70:	learn: 0.0447646	total: 7.79s	remaining: 14.1s
71:	learn: 0.0443706	total: 7.9s	remaining: 14s
72:	learn: 0.0430446	total: 8.01s	remaining: 13.9s
73:	learn: 0.0424090	total: 8.13s	remaining: 13.8s
74:	learn: 0.0412628	total: 8.23s	remaining: 13.7s
75:	learn: 0.0404490	total: 8.34s	remaining: 13.6s
76:	learn: 0.0393314	total: 8.44s	remaining: 13.5s
77:	learn: 0.0386342	total: 8.54s	remaining: 13.4s
78:	learn: 0.0382633	total: 8.64s	remaining: 13.2s
79:	learn: 0.0372343	total: 8.75s	remaining: 13.1s
80:	learn: 0.0367382	total: 8.86s	remaining: 13s
81:	learn: 0.0359115	total: 8.97s	rema

22:	learn: 0.1157694	total: 2.52s	remaining: 19.4s
23:	learn: 0.1134405	total: 2.63s	remaining: 19.3s
24:	learn: 0.1118542	total: 2.74s	remaining: 19.2s
25:	learn: 0.1081731	total: 2.86s	remaining: 19.1s
26:	learn: 0.1058747	total: 2.98s	remaining: 19.1s
27:	learn: 0.1037605	total: 3.09s	remaining: 19s
28:	learn: 0.1013937	total: 3.2s	remaining: 18.9s
29:	learn: 0.0982107	total: 3.3s	remaining: 18.7s
30:	learn: 0.0957780	total: 3.4s	remaining: 18.5s
31:	learn: 0.0929789	total: 3.51s	remaining: 18.4s
32:	learn: 0.0907058	total: 3.62s	remaining: 18.3s
33:	learn: 0.0884027	total: 3.73s	remaining: 18.2s
34:	learn: 0.0856466	total: 3.84s	remaining: 18.1s
35:	learn: 0.0838913	total: 3.96s	remaining: 18s
36:	learn: 0.0822505	total: 4.08s	remaining: 18s
37:	learn: 0.0808910	total: 4.18s	remaining: 17.8s
38:	learn: 0.0789869	total: 4.29s	remaining: 17.7s
39:	learn: 0.0777855	total: 4.39s	remaining: 17.6s
40:	learn: 0.0768464	total: 4.49s	remaining: 17.4s
41:	learn: 0.0746744	total: 4.59s	remain

185:	learn: 0.0063747	total: 20.4s	remaining: 1.54s
186:	learn: 0.0063547	total: 20.5s	remaining: 1.43s
187:	learn: 0.0062910	total: 20.7s	remaining: 1.32s
188:	learn: 0.0062453	total: 20.8s	remaining: 1.21s
189:	learn: 0.0061558	total: 20.9s	remaining: 1.1s
190:	learn: 0.0060870	total: 21s	remaining: 990ms
191:	learn: 0.0060635	total: 21.1s	remaining: 880ms
192:	learn: 0.0060032	total: 21.2s	remaining: 770ms
193:	learn: 0.0059818	total: 21.3s	remaining: 660ms
194:	learn: 0.0059171	total: 21.4s	remaining: 550ms
195:	learn: 0.0058264	total: 21.6s	remaining: 440ms
196:	learn: 0.0057828	total: 21.7s	remaining: 330ms
197:	learn: 0.0057022	total: 21.8s	remaining: 220ms
198:	learn: 0.0056116	total: 21.9s	remaining: 110ms
199:	learn: 0.0055616	total: 22s	remaining: 0us
0:	learn: 0.4962998	total: 106ms	remaining: 21.1s
1:	learn: 0.4013400	total: 204ms	remaining: 20.2s
2:	learn: 0.3321697	total: 307ms	remaining: 20.1s
3:	learn: 0.2847309	total: 419ms	remaining: 20.5s
4:	learn: 0.2543658	total: 

146:	learn: 0.0108708	total: 15.9s	remaining: 5.72s
147:	learn: 0.0107177	total: 16s	remaining: 5.61s
148:	learn: 0.0106209	total: 16.1s	remaining: 5.5s
149:	learn: 0.0104853	total: 16.2s	remaining: 5.39s
150:	learn: 0.0102992	total: 16.3s	remaining: 5.29s
151:	learn: 0.0100796	total: 16.4s	remaining: 5.18s
152:	learn: 0.0098828	total: 16.5s	remaining: 5.08s
153:	learn: 0.0097340	total: 16.6s	remaining: 4.97s
154:	learn: 0.0096056	total: 16.7s	remaining: 4.86s
155:	learn: 0.0095341	total: 16.9s	remaining: 4.75s
156:	learn: 0.0094341	total: 17s	remaining: 4.64s
157:	learn: 0.0093053	total: 17.1s	remaining: 4.53s
158:	learn: 0.0092193	total: 17.2s	remaining: 4.42s
159:	learn: 0.0091073	total: 17.3s	remaining: 4.32s
160:	learn: 0.0090327	total: 17.4s	remaining: 4.21s
161:	learn: 0.0088662	total: 17.5s	remaining: 4.1s
162:	learn: 0.0087299	total: 17.6s	remaining: 3.99s
163:	learn: 0.0086436	total: 17.7s	remaining: 3.88s
164:	learn: 0.0084213	total: 17.8s	remaining: 3.78s
165:	learn: 0.0083

107:	learn: 0.0207669	total: 11.8s	remaining: 10s
108:	learn: 0.0204938	total: 11.9s	remaining: 9.94s
109:	learn: 0.0202716	total: 12s	remaining: 9.82s
110:	learn: 0.0200379	total: 12.1s	remaining: 9.71s
111:	learn: 0.0197775	total: 12.2s	remaining: 9.6s
112:	learn: 0.0195477	total: 12.3s	remaining: 9.49s
113:	learn: 0.0191307	total: 12.4s	remaining: 9.38s
114:	learn: 0.0188507	total: 12.6s	remaining: 9.28s
115:	learn: 0.0186218	total: 12.7s	remaining: 9.17s
116:	learn: 0.0181125	total: 12.8s	remaining: 9.06s
117:	learn: 0.0179875	total: 12.9s	remaining: 8.95s
118:	learn: 0.0176576	total: 13s	remaining: 8.84s
119:	learn: 0.0172238	total: 13.1s	remaining: 8.73s
120:	learn: 0.0170082	total: 13.2s	remaining: 8.62s
121:	learn: 0.0166722	total: 13.3s	remaining: 8.5s
122:	learn: 0.0162338	total: 13.4s	remaining: 8.39s
123:	learn: 0.0159430	total: 13.5s	remaining: 8.28s
124:	learn: 0.0156831	total: 13.6s	remaining: 8.17s
125:	learn: 0.0155725	total: 13.7s	remaining: 8.06s
126:	learn: 0.015370

68:	learn: 0.0415124	total: 7.49s	remaining: 14.2s
69:	learn: 0.0401371	total: 7.61s	remaining: 14.1s
70:	learn: 0.0391290	total: 7.73s	remaining: 14s
71:	learn: 0.0383850	total: 7.83s	remaining: 13.9s
72:	learn: 0.0374189	total: 7.94s	remaining: 13.8s
73:	learn: 0.0364671	total: 8.04s	remaining: 13.7s
74:	learn: 0.0356159	total: 8.16s	remaining: 13.6s
75:	learn: 0.0343830	total: 8.27s	remaining: 13.5s
76:	learn: 0.0338737	total: 8.39s	remaining: 13.4s
77:	learn: 0.0330738	total: 8.49s	remaining: 13.3s
78:	learn: 0.0323850	total: 8.6s	remaining: 13.2s
79:	learn: 0.0321254	total: 8.72s	remaining: 13.1s
80:	learn: 0.0317786	total: 8.82s	remaining: 13s
81:	learn: 0.0313976	total: 8.92s	remaining: 12.8s
82:	learn: 0.0310008	total: 9.02s	remaining: 12.7s
83:	learn: 0.0302583	total: 9.13s	remaining: 12.6s
84:	learn: 0.0298747	total: 9.24s	remaining: 12.5s
85:	learn: 0.0290613	total: 9.35s	remaining: 12.4s
86:	learn: 0.0284647	total: 9.47s	remaining: 12.3s
87:	learn: 0.0275672	total: 9.58s	re

30:	learn: 0.0763827	total: 3.43s	remaining: 18.7s
31:	learn: 0.0724003	total: 3.54s	remaining: 18.6s
32:	learn: 0.0695923	total: 3.65s	remaining: 18.5s
33:	learn: 0.0681653	total: 3.75s	remaining: 18.3s
34:	learn: 0.0665632	total: 3.85s	remaining: 18.2s
35:	learn: 0.0645295	total: 3.96s	remaining: 18s
36:	learn: 0.0625868	total: 4.07s	remaining: 17.9s
37:	learn: 0.0596092	total: 4.17s	remaining: 17.8s
38:	learn: 0.0587998	total: 4.28s	remaining: 17.7s
39:	learn: 0.0573428	total: 4.39s	remaining: 17.6s
40:	learn: 0.0564096	total: 4.5s	remaining: 17.4s
41:	learn: 0.0536639	total: 4.61s	remaining: 17.3s
42:	learn: 0.0525216	total: 4.72s	remaining: 17.2s
43:	learn: 0.0513496	total: 4.82s	remaining: 17.1s
44:	learn: 0.0496708	total: 4.93s	remaining: 17s
45:	learn: 0.0480794	total: 5.04s	remaining: 16.9s
46:	learn: 0.0471828	total: 5.14s	remaining: 16.7s
47:	learn: 0.0458087	total: 5.25s	remaining: 16.6s
48:	learn: 0.0444454	total: 5.37s	remaining: 16.5s
49:	learn: 0.0431554	total: 5.48s	re

192:	learn: 0.0034314	total: 21.9s	remaining: 795ms
193:	learn: 0.0033680	total: 22s	remaining: 681ms
194:	learn: 0.0033176	total: 22.1s	remaining: 568ms
195:	learn: 0.0032766	total: 22.3s	remaining: 454ms
196:	learn: 0.0032559	total: 22.4s	remaining: 341ms
197:	learn: 0.0032256	total: 22.5s	remaining: 227ms
198:	learn: 0.0031954	total: 22.6s	remaining: 114ms
199:	learn: 0.0031616	total: 22.7s	remaining: 0us
0:	learn: 0.4247619	total: 326ms	remaining: 1m 4s
1:	learn: 0.3131321	total: 559ms	remaining: 55.3s
2:	learn: 0.2583678	total: 683ms	remaining: 44.8s
3:	learn: 0.2264319	total: 893ms	remaining: 43.7s
4:	learn: 0.2004348	total: 1.11s	remaining: 43.3s
5:	learn: 0.1869916	total: 1.23s	remaining: 39.9s
6:	learn: 0.1753620	total: 1.37s	remaining: 37.7s
7:	learn: 0.1637531	total: 1.52s	remaining: 36.5s
8:	learn: 0.1529685	total: 1.68s	remaining: 35.7s
9:	learn: 0.1434393	total: 1.8s	remaining: 34.3s
10:	learn: 0.1357870	total: 1.93s	remaining: 33.2s
11:	learn: 0.1323115	total: 2.05s	rema

154:	learn: 0.0042148	total: 18.1s	remaining: 5.25s
155:	learn: 0.0041437	total: 18.2s	remaining: 5.13s
156:	learn: 0.0041359	total: 18.3s	remaining: 5.01s
157:	learn: 0.0040955	total: 18.4s	remaining: 4.9s
158:	learn: 0.0040512	total: 18.5s	remaining: 4.78s
159:	learn: 0.0040036	total: 18.6s	remaining: 4.66s
160:	learn: 0.0039738	total: 18.7s	remaining: 4.54s
161:	learn: 0.0039610	total: 18.8s	remaining: 4.42s
162:	learn: 0.0039174	total: 18.9s	remaining: 4.3s
163:	learn: 0.0038914	total: 19s	remaining: 4.18s
164:	learn: 0.0038453	total: 19.2s	remaining: 4.07s
165:	learn: 0.0038045	total: 19.3s	remaining: 3.96s
166:	learn: 0.0037387	total: 19.5s	remaining: 3.85s
167:	learn: 0.0037289	total: 19.6s	remaining: 3.73s
168:	learn: 0.0037023	total: 19.7s	remaining: 3.62s
169:	learn: 0.0036641	total: 19.8s	remaining: 3.5s
170:	learn: 0.0035897	total: 19.9s	remaining: 3.38s
171:	learn: 0.0035026	total: 20s	remaining: 3.26s
172:	learn: 0.0034613	total: 20.2s	remaining: 3.15s
173:	learn: 0.00343

115:	learn: 0.0078969	total: 13.7s	remaining: 9.93s
116:	learn: 0.0077245	total: 13.8s	remaining: 9.81s
117:	learn: 0.0076339	total: 13.9s	remaining: 9.68s
118:	learn: 0.0075377	total: 14s	remaining: 9.55s
119:	learn: 0.0073480	total: 14.1s	remaining: 9.43s
120:	learn: 0.0072270	total: 14.2s	remaining: 9.3s
121:	learn: 0.0071365	total: 14.3s	remaining: 9.17s
122:	learn: 0.0070541	total: 14.4s	remaining: 9.04s
123:	learn: 0.0068534	total: 14.6s	remaining: 8.92s
124:	learn: 0.0066856	total: 14.7s	remaining: 8.79s
125:	learn: 0.0065758	total: 14.8s	remaining: 8.67s
126:	learn: 0.0064536	total: 14.9s	remaining: 8.55s
127:	learn: 0.0063912	total: 15s	remaining: 8.43s
128:	learn: 0.0062119	total: 15.1s	remaining: 8.3s
129:	learn: 0.0060769	total: 15.2s	remaining: 8.18s
130:	learn: 0.0060131	total: 15.3s	remaining: 8.05s
131:	learn: 0.0059691	total: 15.4s	remaining: 7.93s
132:	learn: 0.0058058	total: 15.5s	remaining: 7.8s
133:	learn: 0.0057418	total: 15.6s	remaining: 7.68s
134:	learn: 0.00567

77:	learn: 0.0190289	total: 8.46s	remaining: 13.2s
78:	learn: 0.0186575	total: 8.56s	remaining: 13.1s
79:	learn: 0.0180193	total: 8.66s	remaining: 13s
80:	learn: 0.0176070	total: 8.77s	remaining: 12.9s
81:	learn: 0.0173571	total: 8.88s	remaining: 12.8s
82:	learn: 0.0170896	total: 8.99s	remaining: 12.7s
83:	learn: 0.0167821	total: 9.1s	remaining: 12.6s
84:	learn: 0.0161840	total: 9.21s	remaining: 12.5s
85:	learn: 0.0156343	total: 9.32s	remaining: 12.4s
86:	learn: 0.0155133	total: 9.42s	remaining: 12.2s
87:	learn: 0.0148713	total: 9.53s	remaining: 12.1s
88:	learn: 0.0145440	total: 9.65s	remaining: 12s
89:	learn: 0.0141152	total: 9.76s	remaining: 11.9s
90:	learn: 0.0137819	total: 9.87s	remaining: 11.8s
91:	learn: 0.0134276	total: 9.98s	remaining: 11.7s
92:	learn: 0.0132899	total: 10.1s	remaining: 11.6s
93:	learn: 0.0129797	total: 10.2s	remaining: 11.5s
94:	learn: 0.0126503	total: 10.3s	remaining: 11.4s
95:	learn: 0.0121959	total: 10.4s	remaining: 11.3s
96:	learn: 0.0120291	total: 10.5s	re

38:	learn: 0.0548562	total: 4.33s	remaining: 17.9s
39:	learn: 0.0531156	total: 4.44s	remaining: 17.7s
40:	learn: 0.0517333	total: 4.54s	remaining: 17.6s
41:	learn: 0.0490530	total: 4.64s	remaining: 17.5s
42:	learn: 0.0477212	total: 4.75s	remaining: 17.3s
43:	learn: 0.0458640	total: 4.85s	remaining: 17.2s
44:	learn: 0.0446542	total: 4.96s	remaining: 17.1s
45:	learn: 0.0442174	total: 5.08s	remaining: 17s
46:	learn: 0.0431102	total: 5.18s	remaining: 16.9s
47:	learn: 0.0418413	total: 5.3s	remaining: 16.8s
48:	learn: 0.0408889	total: 5.41s	remaining: 16.7s
49:	learn: 0.0396913	total: 5.51s	remaining: 16.5s
50:	learn: 0.0379124	total: 5.62s	remaining: 16.4s
51:	learn: 0.0363570	total: 5.72s	remaining: 16.3s
52:	learn: 0.0353985	total: 5.82s	remaining: 16.2s
53:	learn: 0.0349524	total: 5.93s	remaining: 16s
54:	learn: 0.0345995	total: 6.04s	remaining: 15.9s
55:	learn: 0.0329779	total: 6.14s	remaining: 15.8s
56:	learn: 0.0324555	total: 6.26s	remaining: 15.7s
57:	learn: 0.0316506	total: 6.37s	re

198:	learn: 0.0031520	total: 21.8s	remaining: 110ms
199:	learn: 0.0031050	total: 22s	remaining: 0us
0:	learn: 0.4022466	total: 124ms	remaining: 24.7s
1:	learn: 0.2923985	total: 222ms	remaining: 22s
2:	learn: 0.2360720	total: 322ms	remaining: 21.1s
3:	learn: 0.2093207	total: 425ms	remaining: 20.8s
4:	learn: 0.1886435	total: 526ms	remaining: 20.5s
5:	learn: 0.1745071	total: 632ms	remaining: 20.4s
6:	learn: 0.1659436	total: 744ms	remaining: 20.5s
7:	learn: 0.1578034	total: 855ms	remaining: 20.5s
8:	learn: 0.1481475	total: 962ms	remaining: 20.4s
9:	learn: 0.1409233	total: 1.07s	remaining: 20.4s
10:	learn: 0.1331005	total: 1.18s	remaining: 20.3s
11:	learn: 0.1248423	total: 1.28s	remaining: 20.1s
12:	learn: 0.1186133	total: 1.38s	remaining: 19.9s
13:	learn: 0.1137601	total: 1.49s	remaining: 19.8s
14:	learn: 0.1074354	total: 1.59s	remaining: 19.6s
15:	learn: 0.1029003	total: 1.7s	remaining: 19.5s
16:	learn: 0.0997179	total: 1.81s	remaining: 19.5s
17:	learn: 0.0942529	total: 1.92s	remaining: 1

160:	learn: 0.0027825	total: 17.7s	remaining: 4.28s
161:	learn: 0.0027715	total: 17.8s	remaining: 4.17s
162:	learn: 0.0027515	total: 17.9s	remaining: 4.06s
163:	learn: 0.0027126	total: 18s	remaining: 3.95s
164:	learn: 0.0026404	total: 18.1s	remaining: 3.85s
165:	learn: 0.0026081	total: 18.3s	remaining: 3.74s
166:	learn: 0.0025622	total: 18.4s	remaining: 3.63s
167:	learn: 0.0025205	total: 18.5s	remaining: 3.52s
168:	learn: 0.0024801	total: 18.6s	remaining: 3.41s
169:	learn: 0.0024731	total: 18.7s	remaining: 3.3s
170:	learn: 0.0024518	total: 18.8s	remaining: 3.19s
171:	learn: 0.0024365	total: 18.9s	remaining: 3.08s
172:	learn: 0.0024186	total: 19s	remaining: 2.97s
173:	learn: 0.0024046	total: 19.1s	remaining: 2.86s
174:	learn: 0.0023886	total: 19.2s	remaining: 2.75s
175:	learn: 0.0023586	total: 19.3s	remaining: 2.64s
176:	learn: 0.0023467	total: 19.5s	remaining: 2.53s
177:	learn: 0.0023218	total: 19.6s	remaining: 2.42s
178:	learn: 0.0023112	total: 19.7s	remaining: 2.31s
179:	learn: 0.002

122:	learn: 0.0046377	total: 13.7s	remaining: 8.58s
123:	learn: 0.0044840	total: 13.8s	remaining: 8.47s
124:	learn: 0.0044113	total: 13.9s	remaining: 8.36s
125:	learn: 0.0043249	total: 14s	remaining: 8.25s
126:	learn: 0.0043013	total: 14.1s	remaining: 8.13s
127:	learn: 0.0042614	total: 14.3s	remaining: 8.02s
128:	learn: 0.0041545	total: 14.4s	remaining: 7.92s
129:	learn: 0.0040902	total: 14.5s	remaining: 7.8s
130:	learn: 0.0040100	total: 14.6s	remaining: 7.69s
131:	learn: 0.0039457	total: 14.7s	remaining: 7.59s
132:	learn: 0.0038564	total: 14.9s	remaining: 7.48s
133:	learn: 0.0038070	total: 15s	remaining: 7.37s
134:	learn: 0.0037696	total: 15.1s	remaining: 7.26s
135:	learn: 0.0037518	total: 15.2s	remaining: 7.14s
136:	learn: 0.0036850	total: 15.3s	remaining: 7.02s
137:	learn: 0.0036453	total: 15.4s	remaining: 6.91s
138:	learn: 0.0036015	total: 15.5s	remaining: 6.8s
139:	learn: 0.0035054	total: 15.6s	remaining: 6.7s
140:	learn: 0.0034568	total: 15.8s	remaining: 6.59s
141:	learn: 0.00342

84:	learn: 0.0105688	total: 9.1s	remaining: 12.3s
85:	learn: 0.0102407	total: 9.21s	remaining: 12.2s
86:	learn: 0.0099206	total: 9.33s	remaining: 12.1s
87:	learn: 0.0098173	total: 9.44s	remaining: 12s
88:	learn: 0.0096035	total: 9.54s	remaining: 11.9s
89:	learn: 0.0095292	total: 9.64s	remaining: 11.8s
90:	learn: 0.0093660	total: 9.74s	remaining: 11.7s
91:	learn: 0.0091592	total: 9.84s	remaining: 11.6s
92:	learn: 0.0088370	total: 9.95s	remaining: 11.4s
93:	learn: 0.0085886	total: 10.1s	remaining: 11.3s
94:	learn: 0.0085067	total: 10.2s	remaining: 11.2s
95:	learn: 0.0084369	total: 10.3s	remaining: 11.1s
96:	learn: 0.0083255	total: 10.4s	remaining: 11s
97:	learn: 0.0082901	total: 10.5s	remaining: 10.9s
98:	learn: 0.0079899	total: 10.6s	remaining: 10.8s
99:	learn: 0.0077611	total: 10.7s	remaining: 10.7s
100:	learn: 0.0075326	total: 10.8s	remaining: 10.6s
101:	learn: 0.0073947	total: 10.9s	remaining: 10.5s
102:	learn: 0.0071081	total: 11s	remaining: 10.4s
103:	learn: 0.0068695	total: 11.1s	

44:	learn: 0.0324686	total: 4.82s	remaining: 16.6s
45:	learn: 0.0312432	total: 4.92s	remaining: 16.5s
46:	learn: 0.0306278	total: 5.03s	remaining: 16.4s
47:	learn: 0.0294836	total: 5.13s	remaining: 16.2s
48:	learn: 0.0281137	total: 5.23s	remaining: 16.1s
49:	learn: 0.0276436	total: 5.34s	remaining: 16s
50:	learn: 0.0271746	total: 5.44s	remaining: 15.9s
51:	learn: 0.0253344	total: 5.56s	remaining: 15.8s
52:	learn: 0.0245855	total: 5.67s	remaining: 15.7s
53:	learn: 0.0237327	total: 5.78s	remaining: 15.6s
54:	learn: 0.0230112	total: 5.88s	remaining: 15.5s
55:	learn: 0.0225372	total: 5.99s	remaining: 15.4s
56:	learn: 0.0218169	total: 6.09s	remaining: 15.3s
57:	learn: 0.0213965	total: 6.19s	remaining: 15.2s
58:	learn: 0.0205301	total: 6.29s	remaining: 15s
59:	learn: 0.0191933	total: 6.41s	remaining: 15s
60:	learn: 0.0188126	total: 6.51s	remaining: 14.8s
61:	learn: 0.0186421	total: 6.63s	remaining: 14.8s
62:	learn: 0.0183113	total: 6.73s	remaining: 14.6s
63:	learn: 0.0176162	total: 6.84s	rem

4:	learn: 0.1800175	total: 518ms	remaining: 20.2s
5:	learn: 0.1656430	total: 625ms	remaining: 20.2s
6:	learn: 0.1542080	total: 735ms	remaining: 20.3s
7:	learn: 0.1442779	total: 846ms	remaining: 20.3s
8:	learn: 0.1357722	total: 959ms	remaining: 20.4s
9:	learn: 0.1286234	total: 1.06s	remaining: 20.2s
10:	learn: 0.1242607	total: 1.17s	remaining: 20s
11:	learn: 0.1177876	total: 1.27s	remaining: 19.9s
12:	learn: 0.1127169	total: 1.37s	remaining: 19.8s
13:	learn: 0.1088450	total: 1.48s	remaining: 19.6s
14:	learn: 0.1043098	total: 1.58s	remaining: 19.4s
15:	learn: 0.0988065	total: 1.68s	remaining: 19.3s
16:	learn: 0.0932419	total: 1.79s	remaining: 19.3s
17:	learn: 0.0911284	total: 1.9s	remaining: 19.2s
18:	learn: 0.0883333	total: 2.01s	remaining: 19.2s
19:	learn: 0.0870739	total: 2.12s	remaining: 19.1s
20:	learn: 0.0835723	total: 2.23s	remaining: 19s
21:	learn: 0.0788847	total: 2.33s	remaining: 18.9s
22:	learn: 0.0750864	total: 2.44s	remaining: 18.8s
23:	learn: 0.0716764	total: 2.54s	remainin

166:	learn: 0.0025506	total: 17.9s	remaining: 3.54s
167:	learn: 0.0025396	total: 18s	remaining: 3.44s
168:	learn: 0.0025059	total: 18.1s	remaining: 3.33s
169:	learn: 0.0024781	total: 18.2s	remaining: 3.22s
170:	learn: 0.0024296	total: 18.3s	remaining: 3.11s
171:	learn: 0.0024165	total: 18.5s	remaining: 3s
172:	learn: 0.0024030	total: 18.6s	remaining: 2.9s
173:	learn: 0.0023912	total: 18.7s	remaining: 2.79s
174:	learn: 0.0023753	total: 18.8s	remaining: 2.68s
175:	learn: 0.0023431	total: 18.9s	remaining: 2.58s
176:	learn: 0.0023277	total: 19s	remaining: 2.47s
177:	learn: 0.0023029	total: 19.1s	remaining: 2.36s
178:	learn: 0.0022735	total: 19.2s	remaining: 2.25s
179:	learn: 0.0022583	total: 19.3s	remaining: 2.15s
180:	learn: 0.0022076	total: 19.4s	remaining: 2.04s
181:	learn: 0.0021804	total: 19.5s	remaining: 1.93s
182:	learn: 0.0021490	total: 19.6s	remaining: 1.82s
183:	learn: 0.0021231	total: 19.7s	remaining: 1.72s
184:	learn: 0.0021004	total: 19.8s	remaining: 1.61s
185:	learn: 0.002072

128:	learn: 0.0027440	total: 13.9s	remaining: 7.63s
129:	learn: 0.0027193	total: 14s	remaining: 7.52s
130:	learn: 0.0026568	total: 14.1s	remaining: 7.41s
131:	learn: 0.0026313	total: 14.2s	remaining: 7.31s
132:	learn: 0.0026083	total: 14.3s	remaining: 7.2s
133:	learn: 0.0025847	total: 14.4s	remaining: 7.09s
134:	learn: 0.0025772	total: 14.5s	remaining: 6.98s
135:	learn: 0.0025418	total: 14.6s	remaining: 6.87s
136:	learn: 0.0025254	total: 14.7s	remaining: 6.76s
137:	learn: 0.0024872	total: 14.8s	remaining: 6.66s
138:	learn: 0.0024577	total: 14.9s	remaining: 6.55s
139:	learn: 0.0024177	total: 15s	remaining: 6.44s
140:	learn: 0.0023947	total: 15.1s	remaining: 6.33s
141:	learn: 0.0023634	total: 15.2s	remaining: 6.23s
142:	learn: 0.0023363	total: 15.4s	remaining: 6.12s
143:	learn: 0.0022826	total: 15.5s	remaining: 6.01s
144:	learn: 0.0022628	total: 15.6s	remaining: 5.9s
145:	learn: 0.0022385	total: 15.7s	remaining: 5.79s
146:	learn: 0.0022151	total: 15.8s	remaining: 5.68s
147:	learn: 0.0021

91:	learn: 0.0050323	total: 9.88s	remaining: 11.6s
92:	learn: 0.0049519	total: 9.97s	remaining: 11.5s
93:	learn: 0.0048979	total: 10.1s	remaining: 11.4s
94:	learn: 0.0048486	total: 10.2s	remaining: 11.3s
95:	learn: 0.0048055	total: 10.3s	remaining: 11.2s
96:	learn: 0.0047453	total: 10.4s	remaining: 11s
97:	learn: 0.0046076	total: 10.5s	remaining: 10.9s
98:	learn: 0.0044388	total: 10.6s	remaining: 10.8s
99:	learn: 0.0043550	total: 10.7s	remaining: 10.7s
100:	learn: 0.0043043	total: 10.8s	remaining: 10.6s
101:	learn: 0.0042512	total: 10.9s	remaining: 10.5s
102:	learn: 0.0042274	total: 11s	remaining: 10.4s
103:	learn: 0.0041567	total: 11.1s	remaining: 10.3s
104:	learn: 0.0040319	total: 11.3s	remaining: 10.2s
105:	learn: 0.0039638	total: 11.4s	remaining: 10.1s
106:	learn: 0.0039048	total: 11.5s	remaining: 9.96s
107:	learn: 0.0038517	total: 11.6s	remaining: 9.86s
108:	learn: 0.0037870	total: 11.7s	remaining: 9.76s
109:	learn: 0.0037740	total: 11.8s	remaining: 9.65s
110:	learn: 0.0037550	tot

52:	learn: 0.0166646	total: 5.64s	remaining: 15.7s
53:	learn: 0.0163033	total: 5.75s	remaining: 15.6s
54:	learn: 0.0154063	total: 5.86s	remaining: 15.5s
55:	learn: 0.0150380	total: 5.97s	remaining: 15.4s
56:	learn: 0.0146480	total: 6.08s	remaining: 15.2s
57:	learn: 0.0141537	total: 6.18s	remaining: 15.1s
58:	learn: 0.0138071	total: 6.28s	remaining: 15s
59:	learn: 0.0135852	total: 6.38s	remaining: 14.9s
60:	learn: 0.0133462	total: 6.49s	remaining: 14.8s
61:	learn: 0.0127783	total: 6.61s	remaining: 14.7s
62:	learn: 0.0124600	total: 6.72s	remaining: 14.6s
63:	learn: 0.0119812	total: 6.83s	remaining: 14.5s
64:	learn: 0.0115440	total: 6.93s	remaining: 14.4s
65:	learn: 0.0112323	total: 7.04s	remaining: 14.3s
66:	learn: 0.0106629	total: 7.14s	remaining: 14.2s
67:	learn: 0.0104222	total: 7.25s	remaining: 14.1s
68:	learn: 0.0102340	total: 7.35s	remaining: 14s
69:	learn: 0.0098244	total: 7.46s	remaining: 13.9s
70:	learn: 0.0095897	total: 7.57s	remaining: 13.8s
71:	learn: 0.0094686	total: 7.68s	r

12:	learn: 0.1029429	total: 1.39s	remaining: 20.1s
13:	learn: 0.0986109	total: 1.5s	remaining: 19.9s
14:	learn: 0.0922209	total: 1.6s	remaining: 19.7s
15:	learn: 0.0869141	total: 1.7s	remaining: 19.6s
16:	learn: 0.0839478	total: 1.8s	remaining: 19.4s
17:	learn: 0.0776918	total: 1.92s	remaining: 19.4s
18:	learn: 0.0744354	total: 2.02s	remaining: 19.3s
19:	learn: 0.0714454	total: 2.13s	remaining: 19.2s
20:	learn: 0.0681605	total: 2.24s	remaining: 19.1s
21:	learn: 0.0665092	total: 2.35s	remaining: 19s
22:	learn: 0.0623961	total: 2.44s	remaining: 18.8s
23:	learn: 0.0598059	total: 2.57s	remaining: 18.8s
24:	learn: 0.0565405	total: 2.67s	remaining: 18.7s
25:	learn: 0.0545541	total: 2.77s	remaining: 18.5s
26:	learn: 0.0520591	total: 2.87s	remaining: 18.4s
27:	learn: 0.0503634	total: 2.98s	remaining: 18.3s
28:	learn: 0.0469380	total: 3.1s	remaining: 18.3s
29:	learn: 0.0449582	total: 3.21s	remaining: 18.2s
30:	learn: 0.0431055	total: 3.32s	remaining: 18.1s
31:	learn: 0.0424301	total: 3.42s	rema

173:	learn: 0.0017976	total: 18.6s	remaining: 2.77s
174:	learn: 0.0017745	total: 18.7s	remaining: 2.67s
175:	learn: 0.0017645	total: 18.8s	remaining: 2.56s
176:	learn: 0.0017600	total: 18.9s	remaining: 2.45s
177:	learn: 0.0017496	total: 19s	remaining: 2.35s
178:	learn: 0.0017496	total: 19.1s	remaining: 2.24s
179:	learn: 0.0017380	total: 19.2s	remaining: 2.13s
180:	learn: 0.0017082	total: 19.3s	remaining: 2.03s
181:	learn: 0.0016897	total: 19.4s	remaining: 1.92s
182:	learn: 0.0016606	total: 19.5s	remaining: 1.81s
183:	learn: 0.0016460	total: 19.6s	remaining: 1.71s
184:	learn: 0.0016195	total: 19.7s	remaining: 1.6s
185:	learn: 0.0015934	total: 19.8s	remaining: 1.49s
186:	learn: 0.0015746	total: 19.9s	remaining: 1.39s
187:	learn: 0.0015553	total: 20.1s	remaining: 1.28s
188:	learn: 0.0015294	total: 20.2s	remaining: 1.17s
189:	learn: 0.0015211	total: 20.3s	remaining: 1.07s
190:	learn: 0.0015026	total: 20.4s	remaining: 960ms
191:	learn: 0.0014885	total: 20.5s	remaining: 853ms
192:	learn: 0.0

134:	learn: 0.0029256	total: 14.4s	remaining: 6.92s
135:	learn: 0.0029093	total: 14.5s	remaining: 6.82s
136:	learn: 0.0028988	total: 14.6s	remaining: 6.71s
137:	learn: 0.0028787	total: 14.7s	remaining: 6.61s
138:	learn: 0.0028115	total: 14.8s	remaining: 6.5s
139:	learn: 0.0028056	total: 14.9s	remaining: 6.39s
140:	learn: 0.0027899	total: 15s	remaining: 6.28s
141:	learn: 0.0027897	total: 15.1s	remaining: 6.17s
142:	learn: 0.0027897	total: 15.2s	remaining: 6.06s
143:	learn: 0.0027897	total: 15.3s	remaining: 5.96s
144:	learn: 0.0027896	total: 15.4s	remaining: 5.85s
145:	learn: 0.0027557	total: 15.5s	remaining: 5.75s
146:	learn: 0.0027035	total: 15.6s	remaining: 5.64s
147:	learn: 0.0026603	total: 15.8s	remaining: 5.53s
148:	learn: 0.0026398	total: 15.9s	remaining: 5.43s
149:	learn: 0.0025988	total: 16s	remaining: 5.32s
150:	learn: 0.0025775	total: 16.1s	remaining: 5.21s
151:	learn: 0.0025693	total: 16.2s	remaining: 5.1s
152:	learn: 0.0025683	total: 16.3s	remaining: 5s
153:	learn: 0.0025349

95:	learn: 0.0045255	total: 10.3s	remaining: 11.1s
96:	learn: 0.0043699	total: 10.4s	remaining: 11s
97:	learn: 0.0043282	total: 10.5s	remaining: 10.9s
98:	learn: 0.0041522	total: 10.6s	remaining: 10.8s
99:	learn: 0.0040344	total: 10.7s	remaining: 10.7s
100:	learn: 0.0039654	total: 10.8s	remaining: 10.6s
101:	learn: 0.0039015	total: 10.9s	remaining: 10.5s
102:	learn: 0.0038462	total: 11s	remaining: 10.4s
103:	learn: 0.0037665	total: 11.1s	remaining: 10.3s
104:	learn: 0.0036724	total: 11.2s	remaining: 10.2s
105:	learn: 0.0036302	total: 11.3s	remaining: 10s
106:	learn: 0.0035918	total: 11.4s	remaining: 9.94s
107:	learn: 0.0035762	total: 11.5s	remaining: 9.83s
108:	learn: 0.0035485	total: 11.6s	remaining: 9.72s
109:	learn: 0.0034769	total: 11.8s	remaining: 9.61s
110:	learn: 0.0033811	total: 11.9s	remaining: 9.51s
111:	learn: 0.0033069	total: 12s	remaining: 9.4s
112:	learn: 0.0032826	total: 12.1s	remaining: 9.3s
113:	learn: 0.0032108	total: 12.2s	remaining: 9.19s
114:	learn: 0.0031868	total

55:	learn: 0.0117339	total: 5.99s	remaining: 15.4s
56:	learn: 0.0112049	total: 6.1s	remaining: 15.3s
57:	learn: 0.0107953	total: 6.21s	remaining: 15.2s
58:	learn: 0.0103616	total: 6.32s	remaining: 15.1s
59:	learn: 0.0102664	total: 6.44s	remaining: 15s
60:	learn: 0.0100983	total: 6.54s	remaining: 14.9s
61:	learn: 0.0097255	total: 6.65s	remaining: 14.8s
62:	learn: 0.0093741	total: 6.76s	remaining: 14.7s
63:	learn: 0.0091171	total: 6.86s	remaining: 14.6s
64:	learn: 0.0089465	total: 6.97s	remaining: 14.5s
65:	learn: 0.0088377	total: 7.08s	remaining: 14.4s
66:	learn: 0.0087671	total: 7.18s	remaining: 14.3s
67:	learn: 0.0084441	total: 7.29s	remaining: 14.2s
68:	learn: 0.0081870	total: 7.4s	remaining: 14.1s
69:	learn: 0.0078855	total: 7.51s	remaining: 13.9s
70:	learn: 0.0076858	total: 7.61s	remaining: 13.8s
71:	learn: 0.0075687	total: 7.72s	remaining: 13.7s
72:	learn: 0.0074507	total: 7.82s	remaining: 13.6s
73:	learn: 0.0073446	total: 7.92s	remaining: 13.5s
74:	learn: 0.0070834	total: 8.03s	r

16:	learn: 0.0637052	total: 1.82s	remaining: 19.6s
17:	learn: 0.0592902	total: 1.93s	remaining: 19.5s
18:	learn: 0.0578832	total: 2.03s	remaining: 19.3s
19:	learn: 0.0532597	total: 2.13s	remaining: 19.2s
20:	learn: 0.0507350	total: 2.23s	remaining: 19s
21:	learn: 0.0495833	total: 2.35s	remaining: 19s
22:	learn: 0.0450138	total: 2.46s	remaining: 18.9s
23:	learn: 0.0418664	total: 2.56s	remaining: 18.8s
24:	learn: 0.0395161	total: 2.67s	remaining: 18.7s
25:	learn: 0.0363962	total: 2.78s	remaining: 18.6s
26:	learn: 0.0353344	total: 2.88s	remaining: 18.5s
27:	learn: 0.0339104	total: 2.98s	remaining: 18.3s
28:	learn: 0.0327508	total: 3.08s	remaining: 18.2s
29:	learn: 0.0311754	total: 3.19s	remaining: 18.1s
30:	learn: 0.0307376	total: 3.3s	remaining: 18s
31:	learn: 0.0293572	total: 3.41s	remaining: 17.9s
32:	learn: 0.0281754	total: 3.51s	remaining: 17.8s
33:	learn: 0.0274748	total: 3.62s	remaining: 17.7s
34:	learn: 0.0265182	total: 3.73s	remaining: 17.6s
35:	learn: 0.0251525	total: 3.83s	rema

178:	learn: 0.0013547	total: 19s	remaining: 2.23s
179:	learn: 0.0013413	total: 19.2s	remaining: 2.13s
180:	learn: 0.0013247	total: 19.3s	remaining: 2.02s
181:	learn: 0.0013064	total: 19.4s	remaining: 1.92s
182:	learn: 0.0012948	total: 19.5s	remaining: 1.81s
183:	learn: 0.0012847	total: 19.6s	remaining: 1.7s
184:	learn: 0.0012777	total: 19.7s	remaining: 1.6s
185:	learn: 0.0012776	total: 19.8s	remaining: 1.49s
186:	learn: 0.0012567	total: 19.9s	remaining: 1.38s
187:	learn: 0.0012376	total: 20s	remaining: 1.28s
188:	learn: 0.0012353	total: 20.1s	remaining: 1.17s
189:	learn: 0.0012276	total: 20.2s	remaining: 1.06s
190:	learn: 0.0011890	total: 20.3s	remaining: 958ms
191:	learn: 0.0011701	total: 20.4s	remaining: 852ms
192:	learn: 0.0011700	total: 20.6s	remaining: 745ms
193:	learn: 0.0011700	total: 20.7s	remaining: 639ms
194:	learn: 0.0011700	total: 20.8s	remaining: 532ms
195:	learn: 0.0011698	total: 20.9s	remaining: 426ms
196:	learn: 0.0011698	total: 21s	remaining: 319ms
197:	learn: 0.001169

140:	learn: 0.0021056	total: 15s	remaining: 6.29s
141:	learn: 0.0020819	total: 15.1s	remaining: 6.18s
142:	learn: 0.0020407	total: 15.2s	remaining: 6.08s
143:	learn: 0.0019921	total: 15.3s	remaining: 5.97s
144:	learn: 0.0019466	total: 15.4s	remaining: 5.86s
145:	learn: 0.0019007	total: 15.6s	remaining: 5.75s
146:	learn: 0.0018862	total: 15.7s	remaining: 5.65s
147:	learn: 0.0018643	total: 15.8s	remaining: 5.55s
148:	learn: 0.0018521	total: 15.9s	remaining: 5.44s
149:	learn: 0.0018222	total: 16s	remaining: 5.33s
150:	learn: 0.0018016	total: 16.1s	remaining: 5.23s
151:	learn: 0.0017941	total: 16.2s	remaining: 5.12s
152:	learn: 0.0017700	total: 16.3s	remaining: 5.01s
153:	learn: 0.0017458	total: 16.4s	remaining: 4.9s
154:	learn: 0.0017252	total: 16.5s	remaining: 4.8s
155:	learn: 0.0017023	total: 16.6s	remaining: 4.69s
156:	learn: 0.0016911	total: 16.7s	remaining: 4.58s
157:	learn: 0.0016910	total: 16.8s	remaining: 4.48s
158:	learn: 0.0016729	total: 17s	remaining: 4.37s
159:	learn: 0.001660

101:	learn: 0.0035626	total: 10.9s	remaining: 10.5s
102:	learn: 0.0034617	total: 11s	remaining: 10.4s
103:	learn: 0.0033872	total: 11.1s	remaining: 10.3s
104:	learn: 0.0032250	total: 11.2s	remaining: 10.2s
105:	learn: 0.0032136	total: 11.3s	remaining: 10.1s
106:	learn: 0.0031898	total: 11.4s	remaining: 9.94s
107:	learn: 0.0031047	total: 11.5s	remaining: 9.83s
108:	learn: 0.0030312	total: 11.6s	remaining: 9.72s
109:	learn: 0.0029453	total: 11.7s	remaining: 9.61s
110:	learn: 0.0028992	total: 11.8s	remaining: 9.5s
111:	learn: 0.0028692	total: 12s	remaining: 9.39s
112:	learn: 0.0028148	total: 12.1s	remaining: 9.29s
113:	learn: 0.0027576	total: 12.2s	remaining: 9.19s
114:	learn: 0.0027159	total: 12.3s	remaining: 9.08s
115:	learn: 0.0026711	total: 12.4s	remaining: 8.97s
116:	learn: 0.0026462	total: 12.5s	remaining: 8.86s
117:	learn: 0.0026142	total: 12.6s	remaining: 8.75s
118:	learn: 0.0025559	total: 12.7s	remaining: 8.64s
119:	learn: 0.0025089	total: 12.8s	remaining: 8.54s
120:	learn: 0.002

62:	learn: 0.0082664	total: 6.79s	remaining: 14.8s
63:	learn: 0.0081829	total: 6.89s	remaining: 14.6s
64:	learn: 0.0077472	total: 6.99s	remaining: 14.5s
65:	learn: 0.0075240	total: 7.1s	remaining: 14.4s
66:	learn: 0.0074139	total: 7.2s	remaining: 14.3s
67:	learn: 0.0070942	total: 7.31s	remaining: 14.2s
68:	learn: 0.0070075	total: 7.42s	remaining: 14.1s
69:	learn: 0.0069278	total: 7.53s	remaining: 14s
70:	learn: 0.0067359	total: 7.63s	remaining: 13.9s
71:	learn: 0.0065014	total: 7.74s	remaining: 13.8s
72:	learn: 0.0064310	total: 7.85s	remaining: 13.7s
73:	learn: 0.0063605	total: 7.95s	remaining: 13.5s
74:	learn: 0.0060813	total: 8.05s	remaining: 13.4s
75:	learn: 0.0058989	total: 8.15s	remaining: 13.3s
76:	learn: 0.0057380	total: 8.26s	remaining: 13.2s
77:	learn: 0.0056651	total: 8.36s	remaining: 13.1s
78:	learn: 0.0055364	total: 8.47s	remaining: 13s
79:	learn: 0.0054749	total: 8.59s	remaining: 12.9s
80:	learn: 0.0053104	total: 8.69s	remaining: 12.8s
81:	learn: 0.0051918	total: 8.8s	rema

23:	learn: 0.0387792	total: 2.59s	remaining: 19s
24:	learn: 0.0363840	total: 2.7s	remaining: 18.9s
25:	learn: 0.0329509	total: 2.8s	remaining: 18.7s
26:	learn: 0.0318120	total: 2.9s	remaining: 18.6s
27:	learn: 0.0296321	total: 3s	remaining: 18.5s
28:	learn: 0.0263312	total: 3.11s	remaining: 18.3s
29:	learn: 0.0248108	total: 3.22s	remaining: 18.2s
30:	learn: 0.0238237	total: 3.33s	remaining: 18.1s
31:	learn: 0.0229744	total: 3.44s	remaining: 18s
32:	learn: 0.0218565	total: 3.55s	remaining: 18s
33:	learn: 0.0208956	total: 3.66s	remaining: 17.9s
34:	learn: 0.0204056	total: 3.76s	remaining: 17.7s
35:	learn: 0.0190930	total: 3.86s	remaining: 17.6s
36:	learn: 0.0187895	total: 3.96s	remaining: 17.5s
37:	learn: 0.0183480	total: 4.07s	remaining: 17.3s
38:	learn: 0.0175638	total: 4.18s	remaining: 17.3s
39:	learn: 0.0167714	total: 4.29s	remaining: 17.2s
40:	learn: 0.0154318	total: 4.41s	remaining: 17.1s
41:	learn: 0.0145715	total: 4.52s	remaining: 17s
42:	learn: 0.0141379	total: 4.63s	remaining: 

184:	learn: 0.0009267	total: 20.6s	remaining: 1.67s
185:	learn: 0.0009265	total: 20.7s	remaining: 1.56s
186:	learn: 0.0009265	total: 20.8s	remaining: 1.44s
187:	learn: 0.0009265	total: 20.9s	remaining: 1.33s
188:	learn: 0.0009265	total: 21s	remaining: 1.22s
189:	learn: 0.0009240	total: 21.1s	remaining: 1.11s
190:	learn: 0.0009240	total: 21.2s	remaining: 999ms
191:	learn: 0.0009139	total: 21.3s	remaining: 888ms
192:	learn: 0.0008981	total: 21.4s	remaining: 777ms
193:	learn: 0.0008981	total: 21.5s	remaining: 666ms
194:	learn: 0.0008981	total: 21.6s	remaining: 555ms
195:	learn: 0.0008981	total: 21.7s	remaining: 444ms
196:	learn: 0.0008981	total: 21.8s	remaining: 333ms
197:	learn: 0.0008981	total: 21.9s	remaining: 222ms
198:	learn: 0.0008980	total: 22s	remaining: 111ms
199:	learn: 0.0008980	total: 22.2s	remaining: 0us
0:	learn: 0.2825291	total: 109ms	remaining: 21.7s
1:	learn: 0.2180660	total: 210ms	remaining: 20.8s
2:	learn: 0.1861532	total: 311ms	remaining: 20.4s
3:	learn: 0.1627337	tota

146:	learn: 0.0016027	total: 15.6s	remaining: 5.64s
147:	learn: 0.0015830	total: 15.8s	remaining: 5.54s
148:	learn: 0.0015794	total: 15.9s	remaining: 5.43s
149:	learn: 0.0015743	total: 16s	remaining: 5.32s
150:	learn: 0.0015743	total: 16.1s	remaining: 5.21s
151:	learn: 0.0015502	total: 16.2s	remaining: 5.11s
152:	learn: 0.0015148	total: 16.3s	remaining: 5s
153:	learn: 0.0014952	total: 16.4s	remaining: 4.9s
154:	learn: 0.0014630	total: 16.5s	remaining: 4.79s
155:	learn: 0.0014629	total: 16.6s	remaining: 4.68s
156:	learn: 0.0014459	total: 16.7s	remaining: 4.57s
157:	learn: 0.0014459	total: 16.8s	remaining: 4.47s
158:	learn: 0.0014457	total: 16.9s	remaining: 4.36s
159:	learn: 0.0014420	total: 17s	remaining: 4.26s
160:	learn: 0.0014276	total: 17.1s	remaining: 4.15s
161:	learn: 0.0014171	total: 17.2s	remaining: 4.04s
162:	learn: 0.0014011	total: 17.3s	remaining: 3.94s
163:	learn: 0.0013885	total: 17.4s	remaining: 3.83s
164:	learn: 0.0013885	total: 17.5s	remaining: 3.72s
165:	learn: 0.001384

109:	learn: 0.0023999	total: 11.8s	remaining: 9.64s
110:	learn: 0.0023758	total: 11.9s	remaining: 9.53s
111:	learn: 0.0023306	total: 12s	remaining: 9.42s
112:	learn: 0.0022779	total: 12.1s	remaining: 9.31s
113:	learn: 0.0022210	total: 12.2s	remaining: 9.2s
114:	learn: 0.0021936	total: 12.3s	remaining: 9.1s
115:	learn: 0.0021333	total: 12.4s	remaining: 8.99s
116:	learn: 0.0020650	total: 12.5s	remaining: 8.89s
117:	learn: 0.0020209	total: 12.6s	remaining: 8.78s
118:	learn: 0.0020028	total: 12.7s	remaining: 8.67s
119:	learn: 0.0019632	total: 12.8s	remaining: 8.56s
120:	learn: 0.0019093	total: 12.9s	remaining: 8.45s
121:	learn: 0.0019093	total: 13.1s	remaining: 8.34s
122:	learn: 0.0019093	total: 13.2s	remaining: 8.23s
123:	learn: 0.0018838	total: 13.3s	remaining: 8.13s
124:	learn: 0.0018835	total: 13.4s	remaining: 8.03s
125:	learn: 0.0018611	total: 13.5s	remaining: 7.92s
126:	learn: 0.0018417	total: 13.6s	remaining: 7.81s
127:	learn: 0.0018215	total: 13.7s	remaining: 7.71s
128:	learn: 0.00

70:	learn: 0.0057453	total: 7.93s	remaining: 14.4s
71:	learn: 0.0056852	total: 8.04s	remaining: 14.3s
72:	learn: 0.0053811	total: 8.15s	remaining: 14.2s
73:	learn: 0.0053274	total: 8.26s	remaining: 14.1s
74:	learn: 0.0052210	total: 8.37s	remaining: 14s
75:	learn: 0.0050826	total: 8.48s	remaining: 13.8s
76:	learn: 0.0049111	total: 8.59s	remaining: 13.7s
77:	learn: 0.0047221	total: 8.71s	remaining: 13.6s
78:	learn: 0.0045002	total: 8.81s	remaining: 13.5s
79:	learn: 0.0043907	total: 8.92s	remaining: 13.4s
80:	learn: 0.0043503	total: 9.02s	remaining: 13.3s
81:	learn: 0.0043056	total: 9.13s	remaining: 13.1s
82:	learn: 0.0041483	total: 9.25s	remaining: 13s
83:	learn: 0.0040859	total: 9.36s	remaining: 12.9s
84:	learn: 0.0039223	total: 9.47s	remaining: 12.8s
85:	learn: 0.0038438	total: 9.58s	remaining: 12.7s
86:	learn: 0.0038166	total: 9.68s	remaining: 12.6s
87:	learn: 0.0037960	total: 9.78s	remaining: 12.5s
88:	learn: 0.0037357	total: 9.88s	remaining: 12.3s
89:	learn: 0.0036363	total: 9.99s	r

30:	learn: 0.0210450	total: 3.32s	remaining: 18.1s
31:	learn: 0.0207766	total: 3.43s	remaining: 18s
32:	learn: 0.0195076	total: 3.55s	remaining: 18s
33:	learn: 0.0187075	total: 3.65s	remaining: 17.8s
34:	learn: 0.0178582	total: 3.76s	remaining: 17.7s
35:	learn: 0.0168530	total: 3.86s	remaining: 17.6s
36:	learn: 0.0159185	total: 3.96s	remaining: 17.4s
37:	learn: 0.0152047	total: 4.07s	remaining: 17.3s
38:	learn: 0.0146116	total: 4.17s	remaining: 17.2s
39:	learn: 0.0132826	total: 4.29s	remaining: 17.1s
40:	learn: 0.0130775	total: 4.39s	remaining: 17s
41:	learn: 0.0126414	total: 4.51s	remaining: 17s
42:	learn: 0.0121121	total: 4.61s	remaining: 16.8s
43:	learn: 0.0117056	total: 4.71s	remaining: 16.7s
44:	learn: 0.0110937	total: 4.82s	remaining: 16.6s
45:	learn: 0.0105438	total: 4.92s	remaining: 16.5s
46:	learn: 0.0103998	total: 5.02s	remaining: 16.4s
47:	learn: 0.0102016	total: 5.13s	remaining: 16.2s
48:	learn: 0.0100714	total: 5.24s	remaining: 16.1s
49:	learn: 0.0096360	total: 5.35s	remai

191:	learn: 0.0012335	total: 20.6s	remaining: 857ms
192:	learn: 0.0012335	total: 20.7s	remaining: 750ms
193:	learn: 0.0012335	total: 20.8s	remaining: 642ms
194:	learn: 0.0012334	total: 20.9s	remaining: 535ms
195:	learn: 0.0012333	total: 21s	remaining: 428ms
196:	learn: 0.0012333	total: 21.1s	remaining: 321ms
197:	learn: 0.0012333	total: 21.2s	remaining: 214ms
198:	learn: 0.0012332	total: 21.3s	remaining: 107ms
199:	learn: 0.0012332	total: 21.4s	remaining: 0us
0:	learn: 0.2678176	total: 108ms	remaining: 21.5s
1:	learn: 0.2064802	total: 205ms	remaining: 20.3s
2:	learn: 0.1797767	total: 306ms	remaining: 20.1s
3:	learn: 0.1610072	total: 411ms	remaining: 20.1s
4:	learn: 0.1452022	total: 518ms	remaining: 20.2s
5:	learn: 0.1331234	total: 628ms	remaining: 20.3s
6:	learn: 0.1184014	total: 741ms	remaining: 20.4s
7:	learn: 0.1061206	total: 857ms	remaining: 20.6s
8:	learn: 0.0974981	total: 963ms	remaining: 20.4s
9:	learn: 0.0879567	total: 1.07s	remaining: 20.3s
10:	learn: 0.0809197	total: 1.19s	re

153:	learn: 0.0010083	total: 16.5s	remaining: 4.92s
154:	learn: 0.0010083	total: 16.6s	remaining: 4.81s
155:	learn: 0.0009953	total: 16.7s	remaining: 4.71s
156:	learn: 0.0009952	total: 16.8s	remaining: 4.6s
157:	learn: 0.0009816	total: 16.9s	remaining: 4.49s
158:	learn: 0.0009685	total: 17s	remaining: 4.38s
159:	learn: 0.0009685	total: 17.1s	remaining: 4.27s
160:	learn: 0.0009685	total: 17.2s	remaining: 4.17s
161:	learn: 0.0009560	total: 17.3s	remaining: 4.06s
162:	learn: 0.0009464	total: 17.4s	remaining: 3.95s
163:	learn: 0.0009272	total: 17.5s	remaining: 3.85s
164:	learn: 0.0009087	total: 17.6s	remaining: 3.74s
165:	learn: 0.0009087	total: 17.7s	remaining: 3.63s
166:	learn: 0.0009086	total: 17.8s	remaining: 3.53s
167:	learn: 0.0009086	total: 17.9s	remaining: 3.42s
168:	learn: 0.0009086	total: 18.1s	remaining: 3.31s
169:	learn: 0.0009086	total: 18.2s	remaining: 3.21s
170:	learn: 0.0009085	total: 18.3s	remaining: 3.1s
171:	learn: 0.0009084	total: 18.4s	remaining: 2.99s
172:	learn: 0.00

115:	learn: 0.0016603	total: 12.4s	remaining: 8.98s
116:	learn: 0.0016138	total: 12.5s	remaining: 8.87s
117:	learn: 0.0016137	total: 12.6s	remaining: 8.77s
118:	learn: 0.0015816	total: 12.7s	remaining: 8.66s
119:	learn: 0.0015438	total: 12.8s	remaining: 8.56s
120:	learn: 0.0015436	total: 12.9s	remaining: 8.45s
121:	learn: 0.0015436	total: 13.1s	remaining: 8.35s
122:	learn: 0.0015266	total: 13.2s	remaining: 8.24s
123:	learn: 0.0015266	total: 13.3s	remaining: 8.13s
124:	learn: 0.0015265	total: 13.4s	remaining: 8.02s
125:	learn: 0.0015264	total: 13.5s	remaining: 7.91s
126:	learn: 0.0015264	total: 13.6s	remaining: 7.79s
127:	learn: 0.0015264	total: 13.7s	remaining: 7.69s
128:	learn: 0.0015171	total: 13.8s	remaining: 7.58s
129:	learn: 0.0014972	total: 13.9s	remaining: 7.47s
130:	learn: 0.0014809	total: 14s	remaining: 7.37s
131:	learn: 0.0014695	total: 14.1s	remaining: 7.26s
132:	learn: 0.0014542	total: 14.2s	remaining: 7.16s
133:	learn: 0.0014298	total: 14.3s	remaining: 7.05s
134:	learn: 0.

76:	learn: 0.0051746	total: 8.22s	remaining: 13.1s
77:	learn: 0.0048838	total: 8.33s	remaining: 13s
78:	learn: 0.0048573	total: 8.44s	remaining: 12.9s
79:	learn: 0.0046152	total: 8.54s	remaining: 12.8s
80:	learn: 0.0045612	total: 8.65s	remaining: 12.7s
81:	learn: 0.0042749	total: 8.75s	remaining: 12.6s
82:	learn: 0.0042522	total: 8.85s	remaining: 12.5s
83:	learn: 0.0041534	total: 8.96s	remaining: 12.4s
84:	learn: 0.0041200	total: 9.06s	remaining: 12.3s
85:	learn: 0.0039983	total: 9.17s	remaining: 12.2s
86:	learn: 0.0038884	total: 9.29s	remaining: 12.1s
87:	learn: 0.0037441	total: 9.4s	remaining: 12s
88:	learn: 0.0036724	total: 9.51s	remaining: 11.9s
89:	learn: 0.0036075	total: 9.63s	remaining: 11.8s
90:	learn: 0.0035323	total: 9.72s	remaining: 11.6s
91:	learn: 0.0034786	total: 9.83s	remaining: 11.5s
92:	learn: 0.0033151	total: 9.93s	remaining: 11.4s
93:	learn: 0.0032885	total: 10s	remaining: 11.3s
94:	learn: 0.0032544	total: 10.1s	remaining: 11.2s
95:	learn: 0.0031897	total: 10.3s	rema

37:	learn: 0.0193439	total: 4.04s	remaining: 17.2s
38:	learn: 0.0184331	total: 4.14s	remaining: 17.1s
39:	learn: 0.0171818	total: 4.24s	remaining: 17s
40:	learn: 0.0160788	total: 4.34s	remaining: 16.9s
41:	learn: 0.0157481	total: 4.45s	remaining: 16.7s
42:	learn: 0.0147533	total: 4.56s	remaining: 16.7s
43:	learn: 0.0143064	total: 4.67s	remaining: 16.6s
44:	learn: 0.0129514	total: 4.78s	remaining: 16.5s
45:	learn: 0.0124452	total: 4.89s	remaining: 16.4s
46:	learn: 0.0123139	total: 4.99s	remaining: 16.3s
47:	learn: 0.0118708	total: 5.09s	remaining: 16.1s
48:	learn: 0.0114024	total: 5.2s	remaining: 16s
49:	learn: 0.0108777	total: 5.3s	remaining: 15.9s
50:	learn: 0.0104461	total: 5.41s	remaining: 15.8s
51:	learn: 0.0102728	total: 5.51s	remaining: 15.7s
52:	learn: 0.0100792	total: 5.62s	remaining: 15.6s
53:	learn: 0.0099277	total: 5.74s	remaining: 15.5s
54:	learn: 0.0093631	total: 5.85s	remaining: 15.4s
55:	learn: 0.0092075	total: 5.95s	remaining: 15.3s
56:	learn: 0.0088965	total: 6.05s	rem

198:	learn: 0.0008275	total: 21.1s	remaining: 106ms
199:	learn: 0.0008224	total: 21.2s	remaining: 0us
0:	learn: 0.2926202	total: 109ms	remaining: 21.6s
1:	learn: 0.2280297	total: 214ms	remaining: 21.2s
2:	learn: 0.1967422	total: 320ms	remaining: 21s
3:	learn: 0.1745308	total: 424ms	remaining: 20.8s
4:	learn: 0.1567390	total: 526ms	remaining: 20.5s
5:	learn: 0.1412529	total: 635ms	remaining: 20.5s
6:	learn: 0.1283145	total: 737ms	remaining: 20.3s
7:	learn: 0.1184611	total: 838ms	remaining: 20.1s
8:	learn: 0.1076465	total: 951ms	remaining: 20.2s
9:	learn: 0.0996284	total: 1.06s	remaining: 20.1s
10:	learn: 0.0924876	total: 1.17s	remaining: 20.1s
11:	learn: 0.0859122	total: 1.27s	remaining: 20s
12:	learn: 0.0737968	total: 1.39s	remaining: 19.9s
13:	learn: 0.0696076	total: 1.49s	remaining: 19.8s
14:	learn: 0.0664153	total: 1.59s	remaining: 19.7s
15:	learn: 0.0606080	total: 1.7s	remaining: 19.5s
16:	learn: 0.0553402	total: 1.8s	remaining: 19.4s
17:	learn: 0.0511199	total: 1.91s	remaining: 19

160:	learn: 0.0011562	total: 17.1s	remaining: 4.15s
161:	learn: 0.0011562	total: 17.3s	remaining: 4.05s
162:	learn: 0.0011562	total: 17.4s	remaining: 3.94s
163:	learn: 0.0011562	total: 17.5s	remaining: 3.83s
164:	learn: 0.0011562	total: 17.6s	remaining: 3.73s
165:	learn: 0.0011562	total: 17.7s	remaining: 3.62s
166:	learn: 0.0011562	total: 17.8s	remaining: 3.51s
167:	learn: 0.0011562	total: 17.9s	remaining: 3.4s
168:	learn: 0.0011562	total: 18s	remaining: 3.3s
169:	learn: 0.0011371	total: 18.1s	remaining: 3.19s
170:	learn: 0.0011371	total: 18.2s	remaining: 3.09s
171:	learn: 0.0011371	total: 18.3s	remaining: 2.98s
172:	learn: 0.0011371	total: 18.4s	remaining: 2.87s
173:	learn: 0.0011371	total: 18.5s	remaining: 2.77s
174:	learn: 0.0011371	total: 18.6s	remaining: 2.66s
175:	learn: 0.0011370	total: 18.7s	remaining: 2.55s
176:	learn: 0.0011370	total: 18.8s	remaining: 2.44s
177:	learn: 0.0011196	total: 18.9s	remaining: 2.34s
178:	learn: 0.0011196	total: 19s	remaining: 2.23s
179:	learn: 0.0011

121:	learn: 0.0013822	total: 13.1s	remaining: 8.35s
122:	learn: 0.0013045	total: 13.2s	remaining: 8.24s
123:	learn: 0.0012634	total: 13.3s	remaining: 8.13s
124:	learn: 0.0012433	total: 13.4s	remaining: 8.03s
125:	learn: 0.0012083	total: 13.5s	remaining: 7.93s
126:	learn: 0.0012081	total: 13.6s	remaining: 7.82s
127:	learn: 0.0012080	total: 13.7s	remaining: 7.71s
128:	learn: 0.0011975	total: 13.8s	remaining: 7.6s
129:	learn: 0.0011760	total: 13.9s	remaining: 7.49s
130:	learn: 0.0011446	total: 14s	remaining: 7.38s
131:	learn: 0.0011284	total: 14.1s	remaining: 7.28s
132:	learn: 0.0011084	total: 14.2s	remaining: 7.17s
133:	learn: 0.0010855	total: 14.3s	remaining: 7.06s
134:	learn: 0.0010855	total: 14.4s	remaining: 6.96s
135:	learn: 0.0010855	total: 14.6s	remaining: 6.85s
136:	learn: 0.0010854	total: 14.7s	remaining: 6.74s
137:	learn: 0.0010692	total: 14.8s	remaining: 6.63s
138:	learn: 0.0010692	total: 14.9s	remaining: 6.53s
139:	learn: 0.0010599	total: 15s	remaining: 6.42s
140:	learn: 0.001

83:	learn: 0.0026884	total: 9s	remaining: 12.4s
84:	learn: 0.0025891	total: 9.11s	remaining: 12.3s
85:	learn: 0.0025121	total: 9.21s	remaining: 12.2s
86:	learn: 0.0024878	total: 9.31s	remaining: 12.1s
87:	learn: 0.0024182	total: 9.41s	remaining: 12s
88:	learn: 0.0023538	total: 9.52s	remaining: 11.9s
89:	learn: 0.0022819	total: 9.62s	remaining: 11.8s
90:	learn: 0.0022248	total: 9.73s	remaining: 11.7s
91:	learn: 0.0022166	total: 9.84s	remaining: 11.6s
92:	learn: 0.0021821	total: 9.95s	remaining: 11.4s
93:	learn: 0.0021413	total: 10.1s	remaining: 11.3s
94:	learn: 0.0021234	total: 10.2s	remaining: 11.2s
95:	learn: 0.0020891	total: 10.3s	remaining: 11.1s
96:	learn: 0.0020616	total: 10.4s	remaining: 11s
97:	learn: 0.0020255	total: 10.5s	remaining: 10.9s
98:	learn: 0.0019925	total: 10.6s	remaining: 10.8s
99:	learn: 0.0018915	total: 10.7s	remaining: 10.7s
100:	learn: 0.0018496	total: 10.8s	remaining: 10.6s
101:	learn: 0.0018090	total: 10.9s	remaining: 10.5s
102:	learn: 0.0018089	total: 11s	rem

45:	learn: 0.0116696	total: 4.93s	remaining: 16.5s
46:	learn: 0.0112683	total: 5.04s	remaining: 16.4s
47:	learn: 0.0110339	total: 5.15s	remaining: 16.3s
48:	learn: 0.0107042	total: 5.26s	remaining: 16.2s
49:	learn: 0.0102113	total: 5.37s	remaining: 16.1s
50:	learn: 0.0100457	total: 5.47s	remaining: 16s
51:	learn: 0.0099413	total: 5.57s	remaining: 15.9s
52:	learn: 0.0092446	total: 5.67s	remaining: 15.7s
53:	learn: 0.0089621	total: 5.78s	remaining: 15.6s
54:	learn: 0.0087338	total: 5.88s	remaining: 15.5s
55:	learn: 0.0082183	total: 6s	remaining: 15.4s
56:	learn: 0.0078516	total: 6.1s	remaining: 15.3s
57:	learn: 0.0075696	total: 6.21s	remaining: 15.2s
58:	learn: 0.0072441	total: 6.33s	remaining: 15.1s
59:	learn: 0.0069740	total: 6.43s	remaining: 15s
60:	learn: 0.0067424	total: 6.53s	remaining: 14.9s
61:	learn: 0.0060941	total: 6.64s	remaining: 14.8s
62:	learn: 0.0060111	total: 6.74s	remaining: 14.7s
63:	learn: 0.0058973	total: 6.84s	remaining: 14.5s
64:	learn: 0.0056183	total: 6.95s	remai

6:	learn: 0.1141577	total: 778ms	remaining: 21.4s
7:	learn: 0.1058603	total: 884ms	remaining: 21.2s
8:	learn: 0.0969741	total: 986ms	remaining: 20.9s
9:	learn: 0.0894982	total: 1.09s	remaining: 20.7s
10:	learn: 0.0844533	total: 1.19s	remaining: 20.4s
11:	learn: 0.0775351	total: 1.3s	remaining: 20.4s
12:	learn: 0.0679846	total: 1.41s	remaining: 20.3s
13:	learn: 0.0611095	total: 1.52s	remaining: 20.2s
14:	learn: 0.0550419	total: 1.63s	remaining: 20.1s
15:	learn: 0.0505580	total: 1.74s	remaining: 20s
16:	learn: 0.0471836	total: 1.85s	remaining: 19.9s
17:	learn: 0.0418473	total: 1.95s	remaining: 19.7s
18:	learn: 0.0394604	total: 2.05s	remaining: 19.5s
19:	learn: 0.0368128	total: 2.15s	remaining: 19.4s
20:	learn: 0.0345871	total: 2.26s	remaining: 19.2s
21:	learn: 0.0335378	total: 2.37s	remaining: 19.2s
22:	learn: 0.0309443	total: 2.48s	remaining: 19.1s
23:	learn: 0.0287424	total: 2.59s	remaining: 19s
24:	learn: 0.0273259	total: 2.7s	remaining: 18.9s
25:	learn: 0.0251796	total: 2.81s	remaini

167:	learn: 0.0011139	total: 18s	remaining: 3.42s
168:	learn: 0.0011139	total: 18.1s	remaining: 3.32s
169:	learn: 0.0011138	total: 18.2s	remaining: 3.21s
170:	learn: 0.0011138	total: 18.3s	remaining: 3.1s
171:	learn: 0.0011138	total: 18.4s	remaining: 2.99s
172:	learn: 0.0011138	total: 18.5s	remaining: 2.89s
173:	learn: 0.0011138	total: 18.6s	remaining: 2.78s
174:	learn: 0.0011137	total: 18.7s	remaining: 2.67s
175:	learn: 0.0011137	total: 18.8s	remaining: 2.57s
176:	learn: 0.0011137	total: 18.9s	remaining: 2.46s
177:	learn: 0.0011137	total: 19s	remaining: 2.35s
178:	learn: 0.0011137	total: 19.1s	remaining: 2.24s
179:	learn: 0.0011137	total: 19.2s	remaining: 2.14s
180:	learn: 0.0011134	total: 19.3s	remaining: 2.03s
181:	learn: 0.0011134	total: 19.5s	remaining: 1.92s
182:	learn: 0.0011133	total: 19.6s	remaining: 1.82s
183:	learn: 0.0011133	total: 19.7s	remaining: 1.71s
184:	learn: 0.0011133	total: 19.8s	remaining: 1.6s
185:	learn: 0.0011132	total: 19.9s	remaining: 1.5s
186:	learn: 0.00111

{'iterations': 100, 'learning_rate': 0.2}
F1 : 0.563900614864622
Wall time: 22min 9s

Проверим даные на тестовой выборке.

In [61]:
predictions_test = model_LR.predict(features_test)
print("F1 на тестовой выборке:", f1_score(target_test, predictions_test))

F1 на тестовой выборке: 0.6159695817490494


F1 на тестовой выборке: 0.6159695817490494

# Проект для «Викишоп» c TF-IDF

## Подготовка

In [2]:
data = pd.read_csv('datasets/toxic_comments.csv')

In [3]:
data

text  toxic
0       Explanation\nWhy the edits made under my usern...      0
1       D'aww! He matches this background colour I'm s...      0
2       Hey man, I'm really not trying to edit war. It...      0
3       "\nMore\nI can't make any real suggestions on ...      0
4       You, sir, are my hero. Any chance you remember...      0
...                                                   ...    ...
159566  ":::::And for the second time of asking, when ...      0
159567  You should be ashamed of yourself \n\nThat is ...      0
159568  Spitzer \n\nUmm, theres no actual article for ...      0
159569  And it looks like it was actually you who put ...      0
159570  "\nAnd ... I really don't think you understand...      0

[159571 rows x 2 columns]

Импортируем стеммер для английского языка.

In [4]:
stemmer = SnowballStemmer('english')

In [5]:
# Функция возвращает преобразованный стеммингом текст
def stemming(text):
    stemmed = []
    for word in text.split():
        stemmed_word = stemmer.stem(word)
        stemmed.append(stemmed_word)
    stemmed_text = " ".join(stemmed)
    return stemmed_text

In [6]:
stemming(data['text'][0])

"explan whi the edit made under my usernam hardcor metallica fan were reverted? they weren't vandalisms, just closur on some gas after i vote at new york doll fac. and pleas don't remov the templat from the talk page sinc i'm retir now.89.205.38.27"

Применим нашу функцию ко всей колонке `text`.

In [7]:
data['text'] = data['text'].apply(stemming)
data

text  toxic
0       explan whi the edit made under my usernam hard...      0
1       d'aww! he match this background colour i'm see...      0
2       hey man, i'm realli not tri to edit war. it ju...      0
3       " more i can't make ani real suggest on improv...      0
4       you, sir, are my hero. ani chanc you rememb wh...      0
...                                                   ...    ...
159566  ":::::and for the second time of asking, when ...      0
159567  you should be asham of yourself that is a horr...      0
159568  spitzer umm, there no actual articl for prosti...      0
159569  and it look like it was actual you who put on ...      0
159570  " and ... i realli don't think you understand....      0

[159571 rows x 2 columns]

Разобьём датасет на обучающий и тестовый наборы данных.

In [8]:
train, test = train_test_split(data, test_size=0.2, stratify=data['toxic'],random_state=12345)

In [9]:
train.shape[0]

127656

In [10]:
test.shape[0]

31915

In [11]:
# Функция возвращает матрицы TF-IDF для обучающей и тестовой выборок
def tfidfvectors(train, test, column):
    corpus = train[column].values.astype('U')
    nltk.download('stopwords')
    stopwords = set(nltk_stopwords.words('russian'))
    count_tf_idf = TfidfVectorizer(stop_words=stopwords)
    tf_idf_train = count_tf_idf.fit_transform(corpus)
    tf_idf_test = count_tf_idf.transform(test[column])
    return tf_idf_train, tf_idf_test

In [12]:
features_train, features_test = tfidfvectors(train, test, 'text')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\shepelev.ao\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [13]:
features_train

<127656x159008 sparse matrix of type '<class 'numpy.float64'>'
	with 5479656 stored elements in Compressed Sparse Row format>

In [14]:
features_test

<31915x159008 sparse matrix of type '<class 'numpy.float64'>'
	with 1353941 stored elements in Compressed Sparse Row format>

In [15]:
target_train = train['toxic']
target_test = test['toxic']

## Обучение

In [16]:
# Функция возвращает предсказания лучшей модели
def best_model(features_train, 
                  target_train,
                  model,
                  param_grid, 
                  cv, 
                  refit):
    
    my_scorer = {'F1': make_scorer(f1_score)
                #,'MAE': make_scorer(mean_absolute_error)
                }
    
    model = model
    grid = GridSearchCV(model, 
                        param_grid, 
                        cv=cv, 
                        refit=refit, 
                        scoring=my_scorer) 
    grid_result = grid.fit(features_train, target_train)
    best_model = grid_result.best_estimator_
    best_score = grid_result.best_score_
    best_params = grid.best_params_
    print(best_params)
    #print(refit, ": {:.2f}".format(best_score), "%")
    print(refit,":", best_score)
    return(best_model, best_score)

In [22]:
# Сетка параметров для логистической регрессии
PARAM_GRID_LR = {#'penalty': ['l1', 'l2', 'elasticnet', 'none'],
                 'max_iter': [10000],
                 'class_weight': ['balanced'],
                 'solver': ['lbfgs']# ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
                 #,'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
                }
PARAM_GRID_LR;

In [19]:
%%time
model_LR, F1_LR = best_model(features_train, 
                  target_train,
                  LogisticRegression(),
                  PARAM_GRID_LR, 
                  5, 
                  'F1')

C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 531, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1356, in fit
    self.coef_, self.intercept_, n_iter_ = _fit_liblinear(
  File "C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\svm\_base.py", line 930, in _fit_liblinear
    class_weight_ = compute_class_weight(class_weight, classes=classes_,
  File "C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\validation.py", line 72, in inner_f
    return f(**kwargs)
  File "C:\Users\shepelev.ao\Anaconda3\lib\site-packages\sklearn\utils\class_weight.p

{'class_weight': 'balanced', 'max_iter': 10000, 'solver': 'lbfgs'}
F1 : 0.7442439227856392
Wall time: 4min 41s


Проверим метрику на тестовой выборке.

In [20]:
predictions_test = model_LR.predict(features_test)
print("F1 на тестовой выборке:", f1_score(target_test, predictions_test))

F1 на тестовой выборке: 0.7599731363331096


Мы получили удовлетворительный результат.

## Выводы

- Наилучший результат на обучающей и тестовой выборке показала модель логистической регрессии.
- Модель BERT для нашей задачи использовать нецелесообразно, т.к. требуется слишком много времени для построения эмбедингов.
- Выборки в 6000 токенов не достаточно для точного предсказания.
- TF-IDF показала лучшие результаты для нашей задачи, и значительно выиграла в скорости.